--- cinecutie/aboutprefs.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/aboutprefs.h	2010-02-19 22:19:11.838575648 +0100
@@ -22,7 +22,6 @@
 #ifndef ABOUTPREFS_H
 #define ABOUTPREFS_H
 
-#include "guicast.h"
 #include "preferencesthread.h"
 
 class AboutPrefs : public PreferencesDialog
@@ -33,8 +32,7 @@
 	
 
 
-	void create_objects();
-	ArrayList<BC_ListBoxItem*> credits;
+	int create_objects();
 };
 
 
