--- cinecutie/arender.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/arender.C	2010-02-19 22:19:11.838575648 +0100
@@ -26,7 +26,6 @@
 #include "audiodevice.h"
 #include "auto.h"
 #include "autos.h"
-#include "bcsignals.h"
 #include "cache.h"
 #include "condition.h"
 #include "edit.h"
@@ -186,10 +185,8 @@
 	int reconfigure = 0;
 	current_position = input_position;
 
-SET_TRACE
 	while(fragment_position < input_len)
 	{
-SET_TRACE
 		for(int i = 0; i < MAXCHANNELS; i++)
 		{
 			if(buffer_out[i])
@@ -197,35 +194,28 @@
 			else
 				this->audio_out[i] = 0;
 		}
-SET_TRACE
 
 		fragment_len = input_len;
 		if(fragment_position + fragment_len > input_len)
 			fragment_len = input_len - fragment_position;
-SET_TRACE
 
 		reconfigure = vconsole->test_reconfigure(input_position, 
 			fragment_len,
 			last_playback);
-SET_TRACE
 
 //printf("ARender::process_buffer 1 %lld %d\n", input_position, reconfigure);
 
 		if(reconfigure) restart_playback();
-SET_TRACE
 
 		result = process_buffer(fragment_len, input_position);
-SET_TRACE
 
 		fragment_position += fragment_len;
 		input_position += fragment_len;
 		current_position = input_position;
 	}
-SET_TRACE
 
 // Don't delete audio_out on completion
 	bzero(this->audio_out, sizeof(double*) * MAXCHANNELS);
-SET_TRACE
 
 
 
@@ -284,12 +274,11 @@
 {
 	int64_t current_input_length;
 	int reconfigure = 0;
-const int debug = 0;
 
 	first_buffer = 1;
 
 	start_lock->unlock();
-if(debug) printf("ARender::run 1 %d\n", Thread::calculate_realtime());
+//printf("ARender::run 1 %d\n", Thread::calculate_realtime());
 
 	while(!done && !interrupt && !last_playback)
 	{
@@ -297,7 +286,7 @@
 
 		get_boundaries(current_input_length);
 
-if(debug) printf("ARender::run 10 %lld %lld\n", current_position, current_input_length);
+//printf("ARender::run 10 %lld %lld\n", current_position, current_input_length);
 		if(current_input_length)
 		{
 			reconfigure = vconsole->test_reconfigure(current_position, 
@@ -305,7 +294,7 @@
 				last_playback);
 			if(reconfigure) restart_playback();
 		}
-if(debug) printf("ARender::run 20 %lld %lld\n", current_position, current_input_length);
+//printf("ARender::run 20 %lld %lld\n", current_position, current_input_length);
 
 
 // Update tracking if no video is playing.
@@ -328,16 +317,17 @@
 		}
 
 
-if(debug) printf("ARender::run 30 %lld\n", current_input_length);
+
+//printf("ARender::run 30 %lld\n", current_input_length);
 
 
 
 		process_buffer(current_input_length, current_position);
-if(debug) printf("ARender::run 40\n");
+//printf("ARender::run 40\n");
 
 
 		advance_position(get_render_length(current_input_length));
-if(debug) printf("ARender::run 50\n");
+//printf("ARender::run 50\n");
 
 
 		if(vconsole->interrupt) interrupt = 1;
