--- cinecutie/asset.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/asset.h	2010-02-19 22:19:11.848575688 +0100
@@ -34,6 +34,11 @@
 
 #include <stdint.h>
 
+// Time code formats
+#define TC_DROPFRAME 0
+#define TC_NONDROPFRAME 1
+#define TC_PAL 2
+#define TC_FILM 3
 
 // Asset can be one of the following:
 // 1) a pure media file
@@ -64,20 +69,20 @@
 // Load and save parameters for a render dialog
 // Used by render, record, menueffects, preferences
 	void load_defaults(BC_Hash *defaults, 
-		const char *prefix /* = 0 */, 
+		char *prefix /* = 0 */, 
 		int do_format /* = 0 */,
 		int do_compression,
 		int do_path,
 		int do_data_types,
 		int do_bits);
 	void save_defaults(BC_Hash *defaults, 
-		const char *prefix /* = 0 */,
+		char *prefix /* = 0 */,
 		int do_format,     /* Don't save format which is autodetected by file loader */
 		int do_compression,    /* Don't save compression which is fixed by driver */
 		int do_path,
 		int do_data_types,
 		int do_bits);
-	const char* construct_param(const char *param, const char *prefix, char *return_value);
+	char* construct_param(char *param, char *prefix, char *return_value);
 
 
 
@@ -97,13 +102,17 @@
 	int read_video(FileXML *xml);
 	int read_index(FileXML *xml);
 	int reset_index();  // When the index file is wrong, reset the asset values
+
+	int set_timecode(char *tc, int format, int end);
+	int reset_timecode();
+	
 // Output path is the path of the output file if name truncation is desired.
 // It is a "" if; complete names should be used.
 	int write(FileXML *file, 
 		int include_index, 
-		const char *output_path);
+		char *output_path);
 // Write the index data and asset info.  Used by IndexThread.
-	int write_index(const char *path, int data_bytes);
+	int write_index(char *path, int data_bytes);
 
 
 // Necessary for renderfarm to get encoding parameters
@@ -112,9 +121,15 @@
 	int write_index(FileXML *xml);
 	int update_path(char *new_path);
 
+	double total_length_framealigned(double fps);
+
 // Path to file
 	char path[BCTEXTLEN];
 
+// Pipe command
+	char pipe[BCTEXTLEN];
+	int use_pipe;
+
 // Folder in resource manager
 	char folder[BCTEXTLEN];
 
@@ -151,6 +166,14 @@
 	int video_data;       
 	int layers;
 	double frame_rate;
+
+// Timecode information. User setable, in case of errors in source
+	char reel_name[BCTEXTLEN];
+	int reel_number;
+	int64_t tcstart;
+	int64_t tcend;
+	int tcformat;
+
 	int width, height;
 // String or FourCC describing compression
 	char vcodec[BCTEXTLEN];
@@ -193,8 +216,10 @@
 // Set by package render during file creation. -1 means square pixels.
 	double aspect_ratio;
 
-
-
+// for the interlace mode 
+	int interlace_autofixoption;
+	int interlace_mode;
+	int interlace_fixmethod;
 
 // for jpeg compression
 	int jpeg_quality;     
