--- cinecutie/audioesound.C	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/audioesound.C	2010-02-19 22:19:11.848575688 +0100
@@ -105,7 +105,7 @@
 	}
 	esd_in_fd = esd_record_stream_fallback(format, device->in_samplerate, 
 			    	translate_device_string(device->out_config->esound_out_server, device->out_config->esound_out_port), 
-						"Cinelerra");
+						"Cinecutie");
 	return 0;
 }
 
@@ -126,7 +126,7 @@
 	};
 	esd_out_fd = esd_play_stream_fallback(format, device->out_samplerate, 
 			    	translate_device_string(device->out_config->esound_out_server, device->out_config->esound_out_port), 
-						"Cinelerra");
+						"Cinecutie");
 
 	device->device_buffer = esd_get_latency(esd_out);
 	device->device_buffer *= device->out_bits / 8 * device->out_channels;
