--- cinecutie/audiodevice.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/audiodevice.h	2010-02-19 22:19:11.848575688 +0100
@@ -28,7 +28,6 @@
 #include <sys/ioctl.h>
 #include <unistd.h>
 
-#include "audio1394.inc"
 #include "audioalsa.inc"
 #include "audioconfig.inc"
 #include "audiodvb.inc"
@@ -38,6 +37,7 @@
 #include "bctimer.inc"
 #include "binary.h"
 #include "condition.inc"
+#include "dcoffset.inc"
 #include "device1394output.inc"
 #include "maxchannels.h"
 #include "mutex.inc"
@@ -46,8 +46,12 @@
 #include "recordgui.inc"
 #include "sema.inc"
 #include "thread.h"
-#include "vdevice1394.inc"
 #include "videodevice.inc"
+#ifdef HAVE_FIREWIRE
+#include "audio1394.inc"
+#include "device1394output.inc"
+#include "vdevice1394.inc"
+#endif
 
 class AudioLowLevel
 {
