--- cinecutie/autos.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/autos.h	2010-02-19 22:19:11.858573208 +0100
@@ -36,7 +36,8 @@
 class Autos : public List<Auto>
 {
 public:
-	Autos(EDL *edl, Track *track);
+	Autos(EDL *edl, 
+		Track *track);
 		
 	virtual ~Autos();
 
@@ -64,11 +65,19 @@
 
 // Insert keyframe at the point if it doesn't exist
 	Auto* insert_auto(int64_t position);
+// Insert keyframe at the point if it doesn't exist
+// Interpolate it insead of copying
+	Auto* insert_auto_for_editing(int64_t position);
 	void insert_track(Autos *automation, 
 		int64_t start_unit, 
 		int64_t length_units,
 		int replace_default);
 	virtual int load(FileXML *xml);
+	void paste(int64_t start, 
+		int64_t length, 
+		double scale, 
+		FileXML *file, 
+		int default_only);
 	void remove_nonsequential(Auto *keyframe);
 	void optimize();
 
@@ -90,15 +99,9 @@
 // exist.
 	Auto *default_auto;
 
+	int autoidx;
+	int autogrouptype;
 	int type;
-	enum
-	{
-		AUTOMATION_TYPE_FLOAT,
-		AUTOMATION_TYPE_MASK,
-		AUTOMATION_TYPE_INT,
-		AUTOMATION_TYPE_PAN,
-		AUTOMATION_TYPE_PLUGIN
-	};
 
 
 
@@ -112,26 +115,16 @@
 	int clear_all();
 	int insert(int64_t start, int64_t end);
 	int paste_silence(int64_t start, int64_t end);
-// Copy for keyframe clipboard & drag & drop
-// default_only - only copy the default keyframe
 	int copy(int64_t start, 
 		int64_t end, 
 		FileXML *xml, 
 		int default_only,
-		int active_only);
-// Paste for keyframe clipboard & drag & drop
-// default_only - only paste the default keyframe
-	void paste(int64_t start, 
-		int64_t length, 
-		double scale, 
-		FileXML *file, 
-		int default_only,
-		int active_only);
+		int autos_only);
 // Stores the background rendering position in result
 	void clear(int64_t start, 
 		int64_t end, 
 		int shift_autos);
-	virtual void set_automation_mode(int64_t start, int64_t end, int mode);
+	virtual void straighten(int64_t start, int64_t end);
 	int clear_auto(int64_t position);
 	int save(FileXML *xml);
 	virtual int slope_adjustment(int64_t ax, double slope);
