--- cinecutie/avc1394transport.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/avc1394transport.C	2010-02-19 22:19:11.858573208 +0100
@@ -89,7 +89,7 @@
 
 }
 
-void AVC1394Transport::create_objects()
+int AVC1394Transport::create_objects()
 {
 	int x = this->x, y = this->y;
 
@@ -112,6 +112,7 @@
 
 	x_end = x + 10;
 
+	return 0;
 }
 
 int AVC1394Transport::keypress_event(int keypress)
