--- cinecutie/batchrender.h	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/batchrender.h	2010-02-19 22:19:11.858573208 +0100
@@ -170,6 +170,40 @@
 	BatchRenderThread *thread;
 };
 
+
+
+class BatchRenderSaveList : public BC_GenericButton, public Thread
+{
+public:
+	BatchRenderSaveList(BatchRenderThread *thread, 
+			    int x, 
+			    int y);
+	~BatchRenderSaveList();
+	int handle_event();
+	BatchRenderThread *thread;
+	BC_FileBox *gui;
+	void run();
+	virtual int keypress_event();
+	Mutex *startup_lock;
+};
+
+class BatchRenderLoadList : public BC_GenericButton, public Thread
+{
+public:
+	BatchRenderLoadList(BatchRenderThread *thread, 
+			    int x, 
+			    int y);
+	~BatchRenderLoadList();
+	int handle_event();
+	BatchRenderThread *thread;
+	BC_FileBox *gui;
+	void run();
+	virtual int keypress_event();
+	Mutex *startup_lock;
+};
+
+
+
 class BatchRenderList : public BC_ListBox
 {
 public:
@@ -267,6 +301,8 @@
 	BC_Title *list_title;
 	BatchRenderNew *new_batch;
 	BatchRenderDelete *delete_batch;
+	BatchRenderSaveList *savelist_batch;
+	BatchRenderLoadList *loadlist_batch;
 	BatchRenderList *batch_list;
 	BatchRenderStart *start_button;
 	BatchRenderStop *stop_button;
