--- cinecutie/brender.C	2010-02-19 23:44:50.828577566 +0100
+++ /home/user/cinecutie/cinecutie/brender.C	2010-02-19 22:19:11.858573208 +0100
@@ -122,7 +122,7 @@
 	timer->update();
 // Create socket for background process.
 	uuid_t socket_temp;
-	sprintf(socket_path, "/tmp/cinelerra.");
+	sprintf(socket_path, "/tmp/cinecutie.");
 	uuid_generate(socket_temp);
 	uuid_unparse(socket_temp, socket_path + strlen(socket_path));
 SET_TEMP(socket_path);
@@ -191,11 +191,14 @@
 // our position.
 void BRender::restart(EDL *edl)
 {
+//printf("BRender::restart 1\n");
 	BRenderCommand *new_command = new BRenderCommand;
 	map_valid = 0;
 	new_command->copy_edl(edl);
 	new_command->command = BRenderCommand::BRENDER_RESTART;
+//printf("BRender::restart 2\n");
 	thread->send_command(new_command);
+//printf("BRender::restart 3\n");
 // Map should be reallocated before this returns.
 }
 
@@ -436,7 +439,6 @@
 		BRenderCommand *new_command = 0;
 		thread_lock->lock("BRenderThread::run 1");
 
-
 // Got new command
 		if(command_queue)
 		{
@@ -489,16 +491,21 @@
 
 
 			stop();
+//printf("BRenderThread::run 4\n");
 			brender->completion_lock->lock("BRenderThread::run 4");
+//printf("BRenderThread::run 5\n");
 
 			if(new_command->edl->tracks->total_playable_vtracks())
 			{
 				if(command) delete command;
 				command = new_command;
+//printf("BRenderThread::run 6\n");
 				start();
+//printf("BRenderThread::run 7\n");
 			}
 			else
 			{
+//printf("BRenderThread::run 8 %p\n", farm_server);
 				delete new_command;
 				new_command = 0;
 			}
@@ -553,6 +560,7 @@
 		preferences->brender_asset->frame_rate = command->edl->session->frame_rate;
 		preferences->brender_asset->width = command->edl->session->output_w;
 		preferences->brender_asset->height = command->edl->session->output_h;
+		preferences->brender_asset->interlace_mode = command->edl->session->interlace_mode;
 
 // Get last contiguous and reset map.
 // If the framerate changes, last good should be 0 from the user.
