--- cinecutie/browsebutton.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/browsebutton.h	2010-02-19 22:19:11.858573208 +0100
@@ -37,25 +37,27 @@
 		BC_TextBox *textbox, 
 		int x, 
 		int y, 
-		const char *init_directory, 
-		const char *title, 
-		const char *caption, 
-		int want_directory = 0);
+		char *init_directory, 
+		char *title, 
+		char *caption, 
+		int want_directory = 0,
+		const char *recent_prefix = NULL);
 	~BrowseButton();
 	
 	int handle_event();
 	void run();
 	int want_directory;
 	char result[1024];
-	const char *title;
-	const char *caption;
-	const char *init_directory;
+	char *title;
+	char *caption;
+	char *init_directory;
 	BC_TextBox *textbox;
 	MWindow *mwindow;
 	BC_WindowBase *parent_window;
 	BrowseButtonWindow *gui;
 	Mutex *startup_lock;
 	int x, y;
+	const char *recent_prefix;
 };
 
 class BrowseButtonWindow : public BC_FileBox
@@ -64,9 +66,9 @@
 	BrowseButtonWindow(MWindow *mwindow,
 		BrowseButton *button,
 		BC_WindowBase *parent_window, 
-		const char *init_directory, 
-		const char *title, 
-		const char *caption, 
+		char *init_directory, 
+		char *title, 
+		char *caption, 
 		int want_directory);
 	~BrowseButtonWindow();
 };
