--- cinecutie/channeldb.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/channeldb.C	2010-02-19 22:19:11.868576808 +0100
@@ -37,7 +37,7 @@
 	channels.remove_all_objects();
 }
 
-char* ChannelDB::prefix_to_path(char *path, const char *prefix)
+char* ChannelDB::prefix_to_path(char *path, char *prefix)
 {
 	FileSystem fs;
 	char directory[BCTEXTLEN];
@@ -47,7 +47,7 @@
 	return path;
 }
 
-void ChannelDB::load(const char *prefix)
+void ChannelDB::load(char *prefix)
 {
 	FileXML file;
 	char path[BCTEXTLEN];
@@ -71,7 +71,7 @@
 	}
 }
 
-void ChannelDB::save(const char *prefix)
+void ChannelDB::save(char *prefix)
 {
 	char path[BCTEXTLEN];
 	FileXML file;
