--- cinecutie/channelpicker.C	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/channelpicker.C	2010-02-19 22:19:11.868576808 +0100
@@ -466,7 +466,7 @@
 
 
 
-void ChannelPicker::create_objects()
+int ChannelPicker::create_objects()
 {
 	channel_text = 0;
 	update_channel_list();
@@ -480,6 +480,7 @@
 	get_subwindow()->add_subwindow(channel_button = new ChannelButton(this, 
 		x, 
 		y - 1));
+	return 0;
 }
 
 int ChannelPicker::reposition()
@@ -499,7 +500,7 @@
 	return 0;
 }
 
-const char* ChannelPicker::get_source_name(Channel *channel)
+char* ChannelPicker::get_source_name(Channel *channel)
 {
 	if(channel->entry < chanlists[channel->freqtable].count)
 		return chanlists[channel->freqtable].list[channel->entry].name;
@@ -507,7 +508,7 @@
 		return chanlists[channel->freqtable].list[0].name;
 }
 
-const char* ChannelPicker::current_channel_name()
+char* ChannelPicker::current_channel_name()
 {
 	if(get_current_channel_struct())
 	{
