--- cinecutie/clipedit.h	2008-08-03 23:04:26.000000000 +0200
+++ /home/user/cinecutie/cinecutie/clipedit.h	2010-02-19 22:19:11.868576808 +0100
@@ -23,27 +23,22 @@
 #define CLIPEDIT_H
 
 #include "awindow.inc"
-#include "bcdialog.h"
 #include "edl.inc"
 #include "guicast.h"
 #include "mwindow.inc"
+#include "thread.h"
 #include "vwindow.inc"
 
-class ClipEditWindow;
 
-class ClipEdit : public BC_DialogThread
+class ClipEdit : public Thread
 {
 public:
 	ClipEdit(MWindow *mwindow, AWindow *awindow, VWindow *vwindow);
 	~ClipEdit();
 
+	void run();
 	void edit_clip(EDL *clip);
 	void create_clip(EDL *clip);
-// After the window is closed and deleted, this is called.
-	void handle_close_event(int result);
-
-// User creates the window and initializes it here.
-	BC_Window* new_gui();
 
 // If it is being created or edited
 	MWindow *mwindow;
@@ -52,8 +47,6 @@
 
 
 	EDL *clip;
-	EDL *original;
-	ClipEditWindow *window;
 	int create_it;
 };
 
@@ -69,9 +62,12 @@
 	void create_objects();
 
 
+// Use this copy of the pointer in ClipEdit since multiple windows are possible	
+	EDL *clip;
 	int create_it;
 	MWindow *mwindow;
 	ClipEdit *thread;
+	BC_TextBox *titlebox;
 };
 
 
