--- cinecutie/confirmsave.C	2008-10-15 02:20:08.000000000 +0200
+++ /home/user/cinecutie/cinecutie/confirmsave.C	2010-02-19 22:19:11.868576808 +0100
@@ -118,10 +118,9 @@
 }
 
 
-void ConfirmSaveWindow::create_objects()
+int ConfirmSaveWindow::create_objects()
 {
 	int x = 10, y = 10;
-	lock_window("ConfirmSaveWindow::create_objects");
 	add_subwindow(new BC_OKButton(this));
 	add_subwindow(new BC_CancelButton(this));
 
@@ -139,7 +138,7 @@
 	add_subwindow(new BC_OKButton(this));
 	x = get_w() - 100;
 	add_subwindow(new BC_CancelButton(this));
-	unlock_window();
+	return 0;
 }
 
 int ConfirmSaveWindow::resize_event(int w, int h)
