--- cinecutie/cwindow.C	2008-10-15 02:02:26.000000000 +0200
+++ /home/user/cinecutie/cinecutie/cwindow.C	2010-02-19 22:19:11.878577328 +0100
@@ -63,28 +63,21 @@
 	delete playback_cursor;
 }
 
-void CWindow::create_objects()
+int CWindow::create_objects()
 {
 	destination = mwindow->defaults->get("CWINDOW_DESTINATION", 0);
 
-
 	gui = new CWindowGUI(mwindow, this);
-
     gui->create_objects();
 
-
 	playback_engine = new CPlayback(mwindow, this, gui->canvas);
 
-
 // Start command loop
 	playback_engine->create_objects();
-
 	gui->transport->set_engine(playback_engine);
-
 	playback_cursor = new CTracking(mwindow, this);
-
 	playback_cursor->create_objects();
-
+    return 0;
 }
 
 
@@ -153,7 +146,6 @@
 			if(created) *created = 1;
 			if(redraw)
 			{
-// May have to unlock CWindowGUI here.
 				mwindow->gui->lock_window("CWindow::calculate_affected_auto");
 				mwindow->gui->canvas->draw_overlays();
 				mwindow->gui->canvas->flash();
