--- cinecutie/cwindowgui.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/cwindowgui.h	2010-02-19 22:19:11.878577328 +0100
@@ -59,7 +59,7 @@
 	CWindowGUI(MWindow *mwindow, CWindow *cwindow);
 	~CWindowGUI();
 
-    void create_objects();
+    int create_objects();
 	int resize_event(int w, int h);
 
 // Events for the fullscreen canvas fall through to here.
@@ -116,10 +116,8 @@
 	FloatAuto *affected_x;
 	FloatAuto *affected_y;
 	FloatAuto *affected_z;
-// Pointer to mask keyframe being modified in the EDL
-	MaskAuto *mask_keyframe;
-// Copy of original value of mask keyframe for keyframe spanning
-	MaskAuto *orig_mask_keyframe;
+// Keyfrom not affecting transformation being affected
+	Auto *affected_keyframe;
 // Mask point being modified
 	int affected_point;
 // Scrollbar offsets during last button press relative to output
@@ -137,10 +135,6 @@
 	float crop_origin_x1, crop_origin_y1;
 	float crop_origin_x2, crop_origin_y2;
 
-	float ruler_origin_x, ruler_origin_y;
-	int ruler_handle;
-	int ruler_translate;
-
 // Origin for camera and projector operations during last button press
 	float center_x, center_y, center_z;
 	float control_in_x, control_in_y, control_out_x, control_out_y;
@@ -271,7 +265,6 @@
 		int &redraw_canvas, 
 		int &rerender,
 		int do_camera);
-	int do_ruler(int draw, int motion, int button_press, int button_release);
 	int test_zoom(int &redraw);
 	void reset_camera();
 	void reset_projector();
