--- cinecutie/device1394output.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/device1394output.h	2010-02-19 22:19:11.948570683 +0100
@@ -145,15 +145,6 @@
 	AudioDevice *adevice;
 
 	// IOCTL # variables
-	      // dv1394
-   int dv1394_init;
-   int dv1394_shutdown;
-   int dv1394_submit_frames;
-   int dv1394_wait_frames;
-   int dv1394_receive_frames;
-   int dv1394_start_receive;
-   int dv1394_get_status;
-
    // video1394
    int video1394_listen_channel;
    int video1394_unlisten_channel;
