--- cinecutie/device1394input.C	2008-10-19 20:38:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/device1394input.C	2010-02-19 22:19:11.938570804 +0100
@@ -100,7 +100,8 @@
 	}
 }
 
-int Device1394Input::open(int port,
+int Device1394Input::open(const char *path,
+	int port,
 	int channel,
 	int length,
 	int channels,
@@ -125,10 +126,9 @@
 // Initialize grabbing
 	if(fd < 0)
 	{
-#define PATH "/dev/dv1394/0"
-		if((fd = ::open(PATH, O_RDWR)) < 0)
+		if((fd = ::open(path, O_RDWR)) < 0)
 		{
-			printf("Device1394Input::open %s: %s\n", PATH, strerror(errno));
+			printf("Device1394Input::open %s: %s\n", path, strerror(errno));
 		}
 		else
 		{
@@ -178,6 +178,7 @@
 
 
 		audio_buffer = new char[INPUT_SAMPLES * 2 * channels];
+
 		audio_lock = new Condition(0, "Device1394Input::audio_lock");
 		video_lock = new Condition(0, "Device1394Input::video_lock");
 		buffer_lock = new Mutex("Device1394Input::buffer_lock");
@@ -195,7 +196,6 @@
 	{
 // Wait for frame to arrive
 		struct dv1394_status status;
-printf("Device1394Input::run %d done=%d\n", __LINE__, done);
 
 		Thread::enable_cancel();
 		if(ioctl(fd, DV1394_IOC_WAIT_FRAMES, 1))
