--- cinecutie/edit.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/edit.h	2010-02-19 22:19:11.948570683 +0100
@@ -62,7 +62,7 @@
 	int picon_h();
 	double frame_w();
 	double frames_per_picon();
-	int copy(int64_t start, int64_t end, FileXML *xml, const char *output_path);
+	int copy(int64_t start, int64_t end, FileXML *xml, char *output_path);
 // When inherited by a plugin need to resample keyframes
 	virtual void resample(double old_rate, double new_rate) {};
 
