--- cinecutie/edl.inc	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/edl.inc	2010-02-19 22:19:11.948570683 +0100
@@ -60,6 +60,8 @@
 #define ATRANSITION_FOLDER "Audio Transitions"
 #define VTRANSITION_FOLDER "Video Transitions"
 
+#define LABEL_FOLDER "Labels"
+
 // Default folders
 #define CLIP_FOLDER "Clips"
 #define MEDIA_FOLDER "Media"
@@ -78,20 +80,19 @@
 // Active tool in CWindow corresponds to a member of the CPanel array
 // and the current drag operation
 
-#define CPANEL_OPERATIONS 10
+#define CPANEL_OPERATIONS 9
 
 
 #define CWINDOW_NONE -1
 #define CWINDOW_PROTECT     0
 #define CWINDOW_ZOOM        1
 #define CWINDOW_MASK        2
-#define CWINDOW_RULER       3
-#define CWINDOW_CAMERA      4
-#define CWINDOW_PROJECTOR   5
-#define CWINDOW_CROP        6
-#define CWINDOW_EYEDROP     7
-#define CWINDOW_TOOL_WINDOW 8
-#define CWINDOW_TITLESAFE   9
+#define CWINDOW_CAMERA      3
+#define CWINDOW_PROJECTOR   4
+#define CWINDOW_CROP        5
+#define CWINDOW_EYEDROP     6
+#define CWINDOW_TOOL_WINDOW 7
+#define CWINDOW_TITLESAFE   8
 
 // Current dragging operations not available in the CPanel
 #define CWINDOW_SCROLL 0x10
