--- cinecutie/edlsession.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/edlsession.h	2010-02-19 22:19:11.948570683 +0100
@@ -65,6 +65,9 @@
 	void equivalent_output(EDLSession *session, double *result);
 	void dump();
 
+// calculates the frame offset for programme timecode
+	int64_t get_frame_offset();
+
 // Audio
 	int achannel_positions[MAXCHANNELS];
 	AudioOutConfig *aconfig_duplex;
@@ -89,11 +92,10 @@
 	double clipboard_length;
 // Colormodel for intermediate frames
 	int color_model;
+// Interlace Mode for entire project
+	int interlace_mode;
 // Coords for cropping operation
 	int crop_x1, crop_x2, crop_y1, crop_y2;
-	float ruler_x1, ruler_y1;
-	float ruler_x2, ruler_y2;
-// Ruler points relative to the output frame.
 // Current folder in resource window
 	char current_folder[BCTEXTLEN];
 // align cursor on frame boundaries
@@ -139,9 +141,6 @@
 	int labels_follow_edits;
 	int mpeg4_deblock;
 	int plugins_follow_edits;
-// For main menu plugin attaching, 
-// attach 1 standalone on the first track and share it with other tracks
-	int single_standalone;
 	int meter_format;
 	int min_meter_db;
 	int max_meter_db;
@@ -169,6 +168,9 @@
 	int safe_regions;
     int64_t sample_rate;
 	float scrub_speed;
+// Load files as a duration
+	int si_useduration;
+	float si_duration;
 // Show assets in track canvas
 	int show_assets;
 // Show titles in resources
@@ -177,6 +179,8 @@
 	int test_playback_edits;
 // Format to display times in
 	int time_format;
+// Offset for timecode
+	int timecode_offset[4];
 // Format to display nudge in, either seconds or track units.
 	int nudge_seconds;
 // Show tool window in CWindow
