--- cinecutie/filemov.h	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/filemov.h	2010-02-19 22:19:11.958569362 +0100
@@ -92,8 +92,8 @@
 	int64_t get_memory_usage();
 	int colormodel_supported(int colormodel);
 	int can_copy_from(Edit *edit, int64_t position); // This file can copy frames directly from the asset
-	static const char *strtocompression(char *string);
-	static const char *compressiontostr(char *string);
+	static char *strtocompression(char *string);
+	static char *compressiontostr(char *string);
 
 // Fix codec to what AVI or MOV support
 	static void fix_codecs(Asset *asset);
