--- cinecutie/filesndfile.C	2008-10-11 17:46:25.000000000 +0200
+++ /home/user/cinecutie/cinecutie/filesndfile.C	2010-02-19 22:19:11.958569362 +0100
@@ -28,7 +28,7 @@
 #include "filesndfile.h"
 #include "language.h"
 #include "mwindow.inc"
-
+#include "mainerror.h"
 
 FileSndFile::FileSndFile(Asset *asset, File *file)
  : FileBase(asset, file)
@@ -232,19 +232,22 @@
 // Doesn't calculate the length
 			if(fd) format_to_asset();
 		}
+SET_TRACE
 	}
 	else
 	if(wr)
 	{
+printf("FileSNDFile::open 1\n");
 		asset_to_format();
+printf("FileSNDFile::open 1\n");
 		fd = sf_open(asset->path, SFM_WRITE, &fd_config);
+printf("FileSNDFile::open 10 %p\n", fd);
 	}
 
 	if(!fd) 
 	{
 		result = 1;
-		printf("FileSndFile::open_file: ");
-		sf_perror(0);
+		eprintf("%s", sf_strerror(0));
 	}
 
 	return result;
@@ -264,8 +267,7 @@
 // Commented out /* && psf->dataoffset */ in sndfile.c: 761
 	if(sf_seek(fd, sample, SEEK_SET) < 0)
 	{
-		printf("FileSndFile::set_audio_position %lld: failed\n", sample);
-		sf_perror(fd);
+		eprintf("sf_seek() to sample %lld failed, reason: %s\n", sample, sf_strerror(fd));
 		return 1;
 	}
 	return 0;
@@ -278,10 +280,10 @@
 //printf("FileSndFile::read_samples %lld %lld\n", file->current_sample, len);
 // Get temp buffer for interleaved channels
 	if(len <= 0 || len > 1000000)
-		printf("FileSndFile::read_samples len=%d\n", len);
+		eprintf("len=%d\n", len);
 
 	if(!buffer)
-		printf("FileSndFile::read_samples buffer=%p\n", buffer);
+		eprintf("buffer=%p\n", buffer);
 
 	if(temp_allocated && temp_allocated < len)
 	{
@@ -299,7 +301,7 @@
 	result = !sf_read_double(fd, temp_double, len * asset->channels);
 
 	if(result)
-		printf("FileSndFile::read_samples fd=%p temp_double=%p len=%d asset=%p asset->channels=%d\n",
+		eprintf("fd=%p temp_double=%p len=%d asset=%p asset->channels=%d\n",
 			fd, temp_double, len, asset, asset->channels);
 
 // Extract single channel
@@ -383,11 +385,10 @@
 {
 	if(bits_popup) delete bits_popup;
 }
-void SndFileConfig::create_objects()
+int SndFileConfig::create_objects()
 {
 	int x = 10, y = 10;
 
-	lock_window("SndFileConfig::create_objects");
 	bits_popup = 0;
 	switch(asset->format)
 	{
@@ -418,7 +419,7 @@
 		add_subwindow(lohi = new SndFileLOHI(this, x + 170, y));
 	}
 	add_subwindow(new BC_OKButton(this));
-	unlock_window();
+	return 0;
 }
 
 int SndFileConfig::close_event()
