--- cinecutie/filevorbis.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/filevorbis.h	2010-02-19 22:19:11.958569362 +0100
@@ -26,6 +26,7 @@
 #include "filebase.h"
 #include "vorbis/vorbisenc.h"
 #include "vorbis/vorbisfile.h"
+#include "resample.inc"
 
 
 
@@ -47,15 +48,22 @@
 
 	static int check_sig(Asset *asset);
 	int open_file(int rd, int wr);
-	int close_file_derived();
+	int close_file();
 	int write_samples(double **buffer, 
 			int64_t len);
 
 	int read_samples(double *buffer, int64_t len);
-
+	int read_samples_float(float *buffer, int64_t len);
+	int prefer_samples_float();
+	
 // Decoding
 	OggVorbis_File vf;
 	FILE *fd;
+	double **pcm_history;
+	float **pcm_history_float;
+#define HISTORY_MAX 0x100000
+	int history_size;
+	int history_start;
 
 // Encoding
 	vorbis_info vi;
@@ -126,7 +134,7 @@
 	VorbisConfigAudio(BC_WindowBase *parent_window, Asset *asset);
 	~VorbisConfigAudio();
 
-	void create_objects();
+	int create_objects();
 	int close_event();
 
 	VorbisFixedBitrate *fixed_bitrate;
