--- cinecutie/file.inc	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/file.inc	2010-02-19 22:19:11.948570683 +0100
@@ -30,7 +30,6 @@
 #define FILE_UNRECOGNIZED_CODEC 2
 #define FILE_IS_XML 3
 
-// Numeric codes for each file format
 #define FILE_UNKNOWN    		0
 #define FILE_AIFF       		10
 #define FILE_AMPEG      	  	16    // For encoding only
@@ -41,7 +40,6 @@
 #define FILE_AVI_ARNE2          19    // DV type 2
 #define FILE_AVI_AVIFILE        21
 #define FILE_AVI_LAVTOOLS       12    // MJPEG type 1
-#define FILE_FLAC               32
 #define FILE_GIF  	    		7
 #define FILE_JPEG 	    		5
 #define FILE_JPEG_LIST  		8
@@ -63,12 +61,47 @@
 #define FILE_EXR                26
 #define FILE_EXR_LIST           27
 #define FILE_CR2                28
+#define FILE_YUV		29	// mjpegtools YUV4MPEG (aka YUV4MPEG2)
 #define FILE_VORBIS             31
-#define FILE_FFMPEG             33
+#define FILE_RAWDV		32
 
 // For formats supported by plugins, the format number is the plugin number in the 
 // plugin list ORed with 0x8000.
 
+#define FILE_FORMAT_PREFIX(format) ( \
+	format == FILE_AIFF    ? "AIFF" :		\
+	format == FILE_AMPEG   ? "MPEG" :		\
+	format == FILE_AC3     ? "AC3" :		\
+	format == FILE_AU      ? "AU" :			\
+	format == FILE_AVI_ARNE1 ? "AVI_ARNE1" :	\
+	format == FILE_AVI     ? "AVI" :		\
+	format == FILE_AVI_ARNE2 ? "AVI_ARNE2" :	\
+	format == FILE_AVI_AVIFILE ? "AVIFILE" :	\
+	format == FILE_AVI_LAVTOOLS ? "AVI_LAVTOOLS" :	\
+	format == FILE_GIF     ? "GIF" :		\
+	format == FILE_JPEG    ? "JPEG" :		\
+	format == FILE_JPEG_LIST ? "JPEG_LIST" :	\
+	format == FILE_MOV     ? "MOV" :		\
+	format == FILE_MPEG    ? "MPEG" :		\
+	format == FILE_PCM     ? "PCM" :		\
+	format == FILE_PLUGIN  ? "PLUGIN" :		\
+	format == FILE_PNG     ? "PNG" :		\
+	format == FILE_PNG_LIST ? "PNG_LIST" :		\
+	format == FILE_SND     ? "SND" :		\
+	format == FILE_TGA     ? "TGA" :		\
+	format == FILE_TGA_LIST ? "TGA_LIST" :		\
+	format == FILE_TIFF    ? "TIFF" :		\
+	format == FILE_TIFF_LIST ? "TIFF_LIST" :	\
+	format == FILE_VMPEG   ? "VMPEG" :		\
+	format == FILE_VORBIS  ? "VORBIS" :		\
+	format == FILE_WAV     ? "WAV" :		\
+	format == FILE_EXR     ? "EXR" :		\
+	format == FILE_EXR_LIST ? "EXR_LIST" :		\
+	format == FILE_RAWDV   ? "RAWDV" :		\
+	format == FILE_OGG     ? "OGG" :		\
+	format == FILE_YUV     ? "YUV" :		\
+	"UNKNOWN")
+
 #if 0
 N_("AC3")
 N_("Apple/SGI AIFF")
@@ -80,8 +113,7 @@
 N_("AVI DV Type 2")
 N_("AVI Avifile")
 N_("AVI Lavtools")
-N_("FLAC")
-N_("FFMPEG")
+N_("Raw DV")
 N_("JPEG Sequence")
 N_("JPEG")
 N_("Quicktime for Linux")
@@ -99,6 +131,7 @@
 N_("OGG Vorbis")    // For decoding only
 N_("EXR")
 N_("EXR Sequence")
+N_("OGG Theora/Vorbis")
 #endif
 
 #define AC3_NAME "AC3"
@@ -111,6 +144,8 @@
 #define AVI_ARNE2_NAME "AVI DV Type 2"
 #define AVI_AVIFILE_NAME "AVI Avifile"
 #define AVI_LAVTOOLS_NAME "AVI Lavtools"
+#define RAWDV_NAME "Raw DV"
+#define OGG_NAME "OGG Theora/Vorbis"
 #define JPEG_LIST_NAME "JPEG Sequence"
 #define JPEG_NAME "JPEG"
 #define MOV_NAME "Quicktime for Linux"
@@ -126,11 +161,11 @@
 #define TIFF_NAME "TIFF"
 #define TIFF_LIST_NAME "TIFF Sequence"
 #define VMPEG_NAME "MPEG Video"    // For encoding only
-#define CR2_NAME "Raw Camera"
 #define EXR_NAME "EXR"
+#define CR2_NAME "Raw Camera"
 #define EXR_LIST_NAME "EXR Sequence"
-#define FLAC_NAME "FLAC"
-#define FFMPEG_NAME "FFMPEG"
+#define YUV_NAME "YUV4MPEG Stream"
+
 
 #define BITSLINEAR8    8
 #define BITSLINEAR16   16
