--- cinecutie/fileac3.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/fileac3.h	2010-02-19 22:19:11.948570683 +0100
@@ -22,6 +22,9 @@
 #ifndef FILEAC3_H
 #define FILEAC3_H
 
+extern "C" {
+#include <libavcodec/avcodec.h>
+};
 
 #include "filebase.h"
 #include <stdio.h>
@@ -45,10 +48,8 @@
 	int write_samples(double **buffer, int64_t len);
 
 private:
-//	AVCodec *codec;
-//	AVCodecContext *codec_context;
-	void *codec;
-	void *codec_context;
+	AVCodec *codec;
+	AVCodecContext *codec_context;
 	FILE *fd;
 	int16_t *temp_raw;
 	int temp_raw_allocated;
