--- cinecutie/filejpeg.C	2008-10-11 17:46:25.000000000 +0200
+++ /home/user/cinecutie/cinecutie/filejpeg.C	2010-02-19 22:19:11.948570683 +0100
@@ -24,6 +24,7 @@
 #include "edit.h"
 #include "file.h"
 #include "filejpeg.h"
+#include "interlacemodes.h"
 #include "jpegwrapper.h"
 #include "language.h"
 #include "libmjpeg.h"
@@ -31,7 +32,7 @@
 #include "quicktime.h"
 #include "vframe.h"
 #include "videodevice.inc"
-
+#include "mainerror.h"
 
 
 FileJPEG::FileJPEG(Asset *asset, File *file)
@@ -201,7 +202,7 @@
 
 	if(!(stream = fopen(path, "rb")))
 	{
-		perror("FileJPEG::read_frame_header");
+		eprintf("Error while opening \"%s\" for reading. \n%m\n", asset->path);
 		return 1;
 	}
 	
@@ -219,6 +220,8 @@
 	asset->width = jpeg_decompress.image_width;
 	asset->height = jpeg_decompress.image_height;
 
+	asset->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
+
 	jpeg_destroy((j_common_ptr)&jpeg_decompress);
 	fclose(stream);
 	
@@ -245,6 +248,7 @@
 		output->get_color_model(),
 		1);
 
+
 	return 0;
 }
 
@@ -290,10 +294,10 @@
 {
 }
 
-void JPEGConfigVideo::create_objects()
+int JPEGConfigVideo::create_objects()
 {
 	int x = 10, y = 10;
-	lock_window("JPEGConfigVideo::create_objects");
+
 	add_subwindow(new BC_Title(x, y, _("Quality:")));
 	add_subwindow(new BC_ISlider(x + 80, 
 		y,
@@ -308,7 +312,7 @@
 		&asset->jpeg_quality));
 
 	add_subwindow(new BC_OKButton(this));
-	unlock_window();
+	return 0;
 }
 
 int JPEGConfigVideo::close_event()
