--- cinecutie/fileformat.C	2008-10-19 20:38:37.000000000 +0200
+++ /home/user/cinecutie/cinecutie/fileformat.C	2010-02-19 22:19:11.948570683 +0100
@@ -22,8 +22,6 @@
 #include "asset.h"
 #include "assets.h"
 #include "bitspopup.h"
-#include "file.h"
-#include "filesystem.h"
 #include "fileformat.h"
 #include "language.h"
 #include "mwindow.h"
@@ -34,8 +32,8 @@
 
 FileFormat::FileFormat(MWindow *mwindow)
  : BC_Window(PROGRAM_NAME ": File Format", 
-		mwindow->gui->get_abs_cursor_x(0),
-		mwindow->gui->get_abs_cursor_y(0),
+		mwindow->gui->get_abs_cursor_x(1),
+		mwindow->gui->get_abs_cursor_y(1),
  		375, 
 		300, 
 		375, 
@@ -46,7 +44,6 @@
 
 FileFormat::~FileFormat()
 {
-	lock_window("FileFormat::~FileFormat");
 	delete lohi;
 	delete hilo;
 	delete signed_button;
@@ -54,25 +51,20 @@
 	delete rate_button;
 	delete channels_button;
 	delete bitspopup;
-	unlock_window();
 }
 
-void FileFormat::create_objects(Asset *asset, char *string2)
+int FileFormat::create_objects(Asset *asset, char *string2)
 {
 // ================================= copy values
 	this->asset = asset;
 	create_objects_(string2);
 }
 
-void FileFormat::create_objects_(char *string2)
+int FileFormat::create_objects_(char *string2)
 {
-	FileSystem dir;
-	File file;
 	char string[1024];
 	int x1 = 10, x2 = 180;
 	int x = x1, y = 10;
-
-	lock_window("FileFormat::create_objects_");
 	add_subwindow(new BC_Title(x, y, string2));
 	y += 20;
 	add_subwindow(new BC_Title(x, y, _("Assuming raw PCM:")));
@@ -119,7 +111,7 @@
 	
 	add_subwindow(new BC_OKButton(this));
 	add_subwindow(new BC_CancelButton(this));
-	unlock_window();
+	return 0;
 }
 
 FileFormatChannels::FileFormatChannels(int x, int y, FileFormat *fwindow, char *text)
