--- cinecutie/fileogg.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/fileogg.h	2010-02-19 22:19:11.958569362 +0100
@@ -96,6 +96,7 @@
 
 class FileOGG : public FileBase
 {
+friend class PackagingEngineOGG;
 public:
 	FileOGG(Asset *asset, File *file);
 	~FileOGG();
@@ -150,6 +151,7 @@
 
 	int ogg_get_page_of_frame(sync_window_t *sw, long serialno, ogg_page *og, int64_t frame);
 	int ogg_seek_to_keyframe(sync_window_t *sw, long serialno, int64_t frame, int64_t *keyframe_number);
+	int ogg_seek_to_databegin(sync_window_t *sw, long serialno);
 
 
 	int64_t start_sample; // first and last sample inside this file
@@ -174,6 +176,7 @@
 	int64_t ogg_frame_position;    // LAST decoded frame position
 	int64_t next_frame_position;   // what is the next sample read_frames must deliver
 	char theora_keyframe_granule_shift;
+	int final_write;
 };
 
 class OGGConfigAudio;
@@ -235,7 +238,7 @@
 	OGGConfigAudio(BC_WindowBase *parent_window, Asset *asset);
 	~OGGConfigAudio();
 
-	void create_objects();
+	int create_objects();
 	int close_event();
 
 	Asset *asset;
@@ -303,7 +306,7 @@
 	OGGConfigVideo(BC_WindowBase *parent_window, Asset *asset);
 	~OGGConfigVideo();
 
-	void create_objects();
+	int create_objects();
 	int close_event();
 
 	OGGTheoraFixedBitrate *fixed_bitrate;
@@ -313,5 +316,44 @@
 	BC_WindowBase *parent_window;
 };
 
+class PackagingEngineOGG : public PackagingEngine
+{
+public:
+	PackagingEngineOGG();
+	~PackagingEngineOGG();
+	int create_packages_single_farm(
+		EDL *edl,
+		Preferences *preferences,
+		Asset *default_asset, 
+		double total_start, 
+		double total_end);
+	RenderPackage* get_package_single_farm(double frames_per_second, 
+		int client_number,
+		int use_local_rate);
+	int64_t get_progress_max();
+	void get_package_paths(ArrayList<char*> *path_list);
+	int packages_are_done();
+
+private:
+	EDL *edl;
+
+	RenderPackage **packages;
+	int total_packages;
+	double video_package_len;    // Target length of a single package
+
+	Asset *default_asset;
+	Preferences *preferences;
+	int current_package;
+	double total_start;
+	double total_end;
+	int64_t audio_position;
+	int64_t video_position;
+	int64_t audio_start;
+	int64_t video_start;
+	int64_t audio_end;
+	int64_t video_end;
+
+};
+
 
 #endif
