--- cinecutie/floatauto.h	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/floatauto.h	2010-02-19 22:19:11.958569362 +0100
@@ -48,9 +48,16 @@
  	float value_to_percentage();
  	float invalue_to_percentage();
  	float outvalue_to_percentage();
+/* 	float percentage_to_value(float percentage);
+ * 	float percentage_to_invalue(float percentage);
+ * 	float percentage_to_outvalue(float percentage);
+ */
 
 // Control values are relative to value
 	float value, control_in_value, control_out_value;
+// X control positions relative to value position for drawing.
+// In native units of the track.
+	int64_t control_in_position, control_out_position;
 
 private:
 	int value_to_str(char *string, float value);
