--- cinecutie/formatpresets.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/formatpresets.C	2010-02-19 22:19:11.958569362 +0100
@@ -24,7 +24,7 @@
 #include "formatpresets.h"
 #include "new.h"
 #include "setformat.h"
-
+#include "interlacemodes.h"
 
 
 
@@ -60,65 +60,50 @@
 	item = new FormatPresetItem(mwindow, this, _("User Defined"));
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("1080P/60"));
-	item->edl->session->audio_channels = 2;
-	item->edl->session->audio_tracks = 2;
-	item->edl->session->sample_rate = 48000;
-	item->edl->session->video_channels = 1;
-	item->edl->session->video_tracks = 1;
-	item->edl->session->frame_rate = (double)60000.0 / 1001;
-// 	item->edl->session->track_w = 1920;
-// 	item->edl->session->track_h = 1080;
-	item->edl->session->output_w = 1920;
-	item->edl->session->output_h = 1080;
-	item->edl->session->aspect_w = 16;
-	item->edl->session->aspect_h = 9;
-	preset_items.append(item);
-
-	item = new FormatPresetItem(mwindow, this, _("1080P/24"));
+	item = new FormatPresetItem(mwindow, this, _("NTSC"));
 	item->edl->session->audio_channels = 2;
 	item->edl->session->audio_tracks = 2;
 	item->edl->session->sample_rate = 48000;
 	item->edl->session->video_channels = 1;
 	item->edl->session->video_tracks = 1;
-	item->edl->session->frame_rate = 24;
-// 	item->edl->session->track_w = 1920;
-// 	item->edl->session->track_h = 1080;
-	item->edl->session->output_w = 1920;
-	item->edl->session->output_h = 1080;
-	item->edl->session->aspect_w = 16;
-	item->edl->session->aspect_h = 9;
+	item->edl->session->frame_rate = (double)30000.0 / 1001;
+	item->edl->session->output_w = 720;
+	item->edl->session->output_h = 480;
+	item->edl->session->aspect_w = 4;
+	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_BOTTOM_FIRST;
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("1080I"));
+	item = new FormatPresetItem(mwindow, this, _("NTSC Half"));
 	item->edl->session->audio_channels = 2;
 	item->edl->session->audio_tracks = 2;
 	item->edl->session->sample_rate = 48000;
 	item->edl->session->video_channels = 1;
 	item->edl->session->video_tracks = 1;
 	item->edl->session->frame_rate = (double)30000.0 / 1001;
-// 	item->edl->session->track_w = 1920;
-// 	item->edl->session->track_h = 1080;
-	item->edl->session->output_w = 1920;
-	item->edl->session->output_h = 1080;
-	item->edl->session->aspect_w = 16;
-	item->edl->session->aspect_h = 9;
+	item->edl->session->output_w = 360;
+	item->edl->session->output_h = 240;
+	item->edl->session->aspect_w = 4;
+	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("720P/60"));
+	item = new FormatPresetItem(mwindow, this, _("NTSC Progressive"));
 	item->edl->session->audio_channels = 2;
 	item->edl->session->audio_tracks = 2;
 	item->edl->session->sample_rate = 48000;
 	item->edl->session->video_channels = 1;
 	item->edl->session->video_tracks = 1;
 	item->edl->session->frame_rate = (double)60000.0 / 1001;
-	item->edl->session->output_w = 1280;
-	item->edl->session->output_h = 720;
-	item->edl->session->aspect_w = 16;
-	item->edl->session->aspect_h = 9;
+	item->edl->session->output_w = 720;
+	item->edl->session->output_h = 480;
+	item->edl->session->aspect_w = 4;
+	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("576I - DV(D) PAL"));
+
+	item = new FormatPresetItem(mwindow, this, _("PAL"));
   	item->edl->session->audio_channels = 2;
   	item->edl->session->audio_tracks = 2;
   	item->edl->session->sample_rate = 48000;
@@ -129,73 +114,107 @@
 	item->edl->session->output_h = 576;
 	item->edl->session->aspect_w = 4;
 	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_BOTTOM_FIRST;
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("480P - DV(D) NTSC"));
+
+
+	item = new FormatPresetItem(mwindow, this, _("PAL Half"));
 	item->edl->session->audio_channels = 2;
 	item->edl->session->audio_tracks = 2;
 	item->edl->session->sample_rate = 48000;
 	item->edl->session->video_channels = 1;
 	item->edl->session->video_tracks = 1;
-	item->edl->session->frame_rate = (double)60000.0 / 1001;
+	item->edl->session->frame_rate = 25;
+	item->edl->session->output_w = 360;
+	item->edl->session->output_h = 288;
+	item->edl->session->aspect_w = 4;
+	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
+	preset_items.append(item);
+
+
+
+	item = new FormatPresetItem(mwindow, this, _("PAL Progressive"));
+  	item->edl->session->audio_channels = 2;
+  	item->edl->session->audio_tracks = 2;
+  	item->edl->session->sample_rate = 48000;
+  	item->edl->session->video_channels = 1;
+  	item->edl->session->video_tracks = 1;
+	item->edl->session->frame_rate = 50;
 	item->edl->session->output_w = 720;
-	item->edl->session->output_h = 480;
+	item->edl->session->output_h = 576;
 	item->edl->session->aspect_w = 4;
 	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("480I - DV(D) NTSC"));
+
+	item = new FormatPresetItem(mwindow, this, _("1080P/60"));
 	item->edl->session->audio_channels = 2;
 	item->edl->session->audio_tracks = 2;
 	item->edl->session->sample_rate = 48000;
 	item->edl->session->video_channels = 1;
 	item->edl->session->video_tracks = 1;
-	item->edl->session->frame_rate = (double)30000.0 / 1001;
-	item->edl->session->output_w = 720;
-	item->edl->session->output_h = 480;
-	item->edl->session->aspect_w = 4;
-	item->edl->session->aspect_h = 3;
+	item->edl->session->frame_rate = (double)60000.0 / 1001;
+// 	item->edl->session->track_w = 1920;
+// 	item->edl->session->track_h = 1080;
+	item->edl->session->output_w = 1920;
+	item->edl->session->output_h = 1080;
+	item->edl->session->aspect_w = 16;
+	item->edl->session->aspect_h = 9;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("YouTube"));
-	item->edl->session->audio_channels = 1;
-	item->edl->session->audio_tracks = 1;
+	item = new FormatPresetItem(mwindow, this, _("1080P/24"));
+	item->edl->session->audio_channels = 2;
+	item->edl->session->audio_tracks = 2;
 	item->edl->session->sample_rate = 48000;
 	item->edl->session->video_channels = 1;
 	item->edl->session->video_tracks = 1;
-	item->edl->session->frame_rate = (double)30000.0 / 1001;
-	item->edl->session->output_w = 424;
-	item->edl->session->output_h = 318;
-	item->edl->session->aspect_w = 4;
-	item->edl->session->aspect_h = 3;
+	item->edl->session->frame_rate = 24;
+// 	item->edl->session->track_w = 1920;
+// 	item->edl->session->track_h = 1080;
+	item->edl->session->output_w = 1920;
+	item->edl->session->output_h = 1080;
+	item->edl->session->aspect_w = 16;
+	item->edl->session->aspect_h = 9;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("Half D-1 PAL"));
+	item = new FormatPresetItem(mwindow, this, _("1080I"));
 	item->edl->session->audio_channels = 2;
 	item->edl->session->audio_tracks = 2;
 	item->edl->session->sample_rate = 48000;
 	item->edl->session->video_channels = 1;
 	item->edl->session->video_tracks = 1;
-	item->edl->session->frame_rate = 25;
-	item->edl->session->output_w = 360;
-	item->edl->session->output_h = 288;
-	item->edl->session->aspect_w = 4;
-	item->edl->session->aspect_h = 3;
+	item->edl->session->frame_rate = (double)30000.0 / 1001;
+// 	item->edl->session->track_w = 1920;
+// 	item->edl->session->track_h = 1080;
+	item->edl->session->output_w = 1920;
+	item->edl->session->output_h = 1080;
+	item->edl->session->aspect_w = 16;
+	item->edl->session->aspect_h = 9;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_BOTTOM_FIRST;
 	preset_items.append(item);
 
-	item = new FormatPresetItem(mwindow, this, _("Half D-1 NTSC"));
+	item = new FormatPresetItem(mwindow, this, _("720P/60"));
 	item->edl->session->audio_channels = 2;
 	item->edl->session->audio_tracks = 2;
 	item->edl->session->sample_rate = 48000;
 	item->edl->session->video_channels = 1;
 	item->edl->session->video_tracks = 1;
-	item->edl->session->frame_rate = (double)30000.0 / 1001;
-	item->edl->session->output_w = 360;
-	item->edl->session->output_h = 240;
-	item->edl->session->aspect_w = 4;
-	item->edl->session->aspect_h = 3;
+	item->edl->session->frame_rate = (double)60000.0 / 1001;
+	item->edl->session->output_w = 1280;
+	item->edl->session->output_h = 720;
+	item->edl->session->aspect_w = 16;
+	item->edl->session->aspect_h = 9;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;	
 	preset_items.append(item);
 
+
+
+
 	item = new FormatPresetItem(mwindow, this, _("Internet"));
 	item->edl->session->audio_channels = 1;
 	item->edl->session->audio_tracks = 1;
@@ -207,6 +226,7 @@
 	item->edl->session->output_h = 240;
 	item->edl->session->aspect_w = 4;
 	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
 	preset_items.append(item);
 
 	item = new FormatPresetItem(mwindow, this, _("CD Audio"));
@@ -220,6 +240,7 @@
 	item->edl->session->output_h = 480;
 	item->edl->session->aspect_w = 4;
 	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
 	preset_items.append(item);
 
 	item = new FormatPresetItem(mwindow, this, _("DAT Audio"));
@@ -233,6 +254,7 @@
 	item->edl->session->output_h = 480;
 	item->edl->session->aspect_w = 4;
 	item->edl->session->aspect_h = 3;
+	item->edl->session->interlace_mode = BC_ILACE_MODE_NOTINTERLACED;
 	preset_items.append(item);
 
 
@@ -267,7 +289,8 @@
 			edl->session->output_w == preset->edl->session->output_w &&
 			edl->session->output_h == preset->edl->session->output_h &&
 			edl->session->aspect_w == preset->edl->session->aspect_w &&
-			edl->session->aspect_h == preset->edl->session->aspect_h)
+			edl->session->aspect_h == preset->edl->session->aspect_h && 
+			edl->session->interlace_mode == preset->edl->session->interlace_mode)
 		{
 			return preset;
 		}
@@ -275,13 +298,13 @@
 	return 0;
 }
 
-const char* FormatPresets::get_preset_text(EDL *edl)
+char* FormatPresets::get_preset_text(EDL *edl)
 {
 	FormatPresetItem *item = find_preset(edl);
 	if(item) 
 		return item->get_text();
 	else
-		return "Custom";
+		return "User Defined";
 }
 
 
