--- cinecutie/framecache.C	2008-07-26 03:56:57.000000000 +0200
+++ /home/user/cinecutie/cinecutie/framecache.C	2010-02-19 22:19:11.958569362 +0100
@@ -95,34 +95,8 @@
 	{
 		if(result->data) 
 		{
-// Frame may have come from the readahead thread.
-// Those frames are in the codec color model.
-// But to pass frame_exists, they must be identical.
-// 			cmodel_transfer(frame->get_rows(), 
-// 				result->data->get_rows(),
-// 				result->data->get_y(),
-// 				result->data->get_u(),
-// 				result->data->get_v(),
-// 				frame->get_y(),
-// 				frame->get_u(),
-// 				frame->get_v(),
-// 				0, 
-// 				0, 
-// 				result->data->get_w(), 
-// 				result->data->get_h(),
-// 				0, 
-// 				0, 
-// 				frame->get_w(), 
-// 				frame->get_h(),
-// 				result->data->get_color_model(), 
-// 				frame->get_color_model(),
-// 				0,
-// 				result->data->get_w(),
-// 				frame->get_w());
 			frame->copy_from(result->data);
-// This would have copied the color matrix for interpolate, but
-// required the same plugin stack as the reader.
-//			frame->copy_stacks(result->data);
+			frame->copy_stacks(result->data);
 		}
 		result->age = get_age();
 	}
@@ -228,24 +202,9 @@
 	FrameCacheItem *item = (FrameCacheItem*)get_item(position);
 	while(item && item->position == position)
 	{
-// printf("FrameCache::frame_exists %d %f,%f %d,%d %d,%d format match=%d item->data=%p\n",
-// __LINE__,
-// item->frame_rate,
-// frame_rate,
-// item->layer,
-// layer,
-// item->asset_id,
-// asset_id,
-// format->equivalent(item->data, 1),
-// item->data);
-// format->dump_params();
-
-// This originally tested the frame stacks because a change in the 
-// interpolate plugin could cause CR2 to interpolate or not interpolate.
-// This was disabled.
 		if(EQUIV(item->frame_rate, frame_rate) &&
 			layer == item->layer &&
-			format->equivalent(item->data, 0) &&
+			format->equivalent(item->data, 1) &&
 			(asset_id == -1 || item->asset_id == -1 || asset_id == item->asset_id))
 		{
 			*item_return = item;
@@ -269,19 +228,6 @@
 	FrameCacheItem *item = (FrameCacheItem*)get_item(position);
 	while(item && item->position == position)
 	{
-// printf("FrameCache::frame_exists %d %f,%f %d,%d %d,%d %d,%d\n",
-// __LINE__,
-// item->frame_rate,
-// frame_rate,
-// item->layer,
-// layer,
-// item->data->get_color_model(),
-// color_model,
-// item->data->get_w(),
-// w,
-// item->data->get_h(),
-// h);
-
 		if(EQUIV(item->frame_rate, frame_rate) &&
 			layer == item->layer &&
 			color_model == item->data->get_color_model() &&
