--- cinecutie/interfaceprefs.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/interfaceprefs.C	2010-02-19 22:19:11.968571001 +0100
@@ -46,7 +46,7 @@
 {
 }
 
-void InterfacePrefs::create_objects()
+int InterfacePrefs::create_objects()
 {
 	int y, x, value;
 	BC_Resources *resources = BC_WindowBase::get_resources();
@@ -169,7 +169,7 @@
 	add_subwindow(thumbnails = new ViewThumbnails(x, y, pwindow));
 
 	y += 35;
-	add_subwindow(new BC_Title(x, y, _("Clicking on edit boundaries does what:")));
+	add_subwindow(new BC_Title(x, y, _("Dragging edit boundaries does what:")));
 	y += 25;
 	add_subwindow(new BC_Title(x, y, _("Button 1:")));
 	
@@ -212,16 +212,17 @@
 	add_subwindow(max_db = new MeterMaxDB(pwindow, string, x, y));
 
 	x = x1;
-	y += 30;
+	y += 20;
 	ViewTheme *theme;
 	add_subwindow(new BC_Title(x, y, _("Theme:")));
 	x += 60;
 	add_subwindow(theme = new ViewTheme(x, y, pwindow));
 	theme->create_objects();
 
+	return 0;
 }
 
-const char* InterfacePrefs::behavior_to_text(int mode)
+char* InterfacePrefs::behavior_to_text(int mode)
 {
 	switch(mode)
 	{
@@ -443,7 +444,7 @@
 
 ViewBehaviourText::ViewBehaviourText(int x, 
 	int y, 
-	const char *text, 
+	char *text, 
 	PreferencesWindow *pwindow, 
 	int *output)
  : BC_PopupMenu(x, y, 200, text)
@@ -459,13 +460,14 @@
 {
 }
 
-void ViewBehaviourText::create_objects()
+int ViewBehaviourText::create_objects()
 {
 // Video4linux versions are automatically detected
 	add_item(new ViewBehaviourItem(this, _(MOVE_ALL_EDITS_TITLE), MOVE_ALL_EDITS));
 	add_item(new ViewBehaviourItem(this, _(MOVE_ONE_EDIT_TITLE), MOVE_ONE_EDIT));
 	add_item(new ViewBehaviourItem(this, _(MOVE_NO_EDITS_TITLE), MOVE_NO_EDITS));
 	add_item(new ViewBehaviourItem(this, _(MOVE_EDITS_DISABLED_TITLE), MOVE_EDITS_DISABLED));
+	return 0;
 }
 
 
