--- cinecutie/mainclock.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/mainclock.C	2010-02-19 22:19:11.968571001 +0100
@@ -36,15 +36,24 @@
 		w)
 {
 	this->mwindow = mwindow;
+	position_offset = 0;
 }
 
 MainClock::~MainClock()
 {
 }
 
+void MainClock::set_frame_offset(double value)
+{
+	position_offset = value;
+}
+
 void MainClock::update(double position)
 {
 	char string[BCTEXTLEN];
+
+	position += position_offset;
+
 	Units::totext(string, 
 		position,
 		mwindow->edl->session->time_format,
