--- cinecutie/mainsession.inc	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/mainsession.inc	2010-02-19 22:19:11.968571001 +0100
@@ -47,17 +47,17 @@
 	DRAG_PAN,          
 	DRAG_PLAY,         
 	DRAG_MUTE,         
-	DRAG_CAMERA_X,      // 20
+	DRAG_CAMERA_X,       
 	DRAG_CAMERA_Y,       
 	DRAG_CAMERA_Z,       
-	DRAG_MASK_PRE,
+	DRAG_MASK_PRE,      // 20
 	DRAG_MASK,
 	DRAG_MODE_PRE,         
 	DRAG_MODE,         
 	DRAG_PROJECTOR_X,    
 	DRAG_PROJECTOR_Y,    
 	DRAG_PROJECTOR_Z,    
-	DRAG_PLUGINKEY_PRE,    // 30
+	DRAG_PLUGINKEY_PRE,    
 	DRAG_PLUGINKEY,    
 	DRAG_PLUGINHANDLE1,        // Waiting to move out of drag threshold
 	DRAG_PLUGINHANDLE2,         // Dragging outside drag threshold
