--- cinecutie/main.C	2008-08-06 20:45:37.000000000 +0200
+++ /home/user/cinecutie/cinecutie/main.C	2010-02-19 22:19:11.968571001 +0100
@@ -35,13 +35,16 @@
 #include "pluginserver.h"
 #include "preferences.h"
 #include "renderfarmclient.h"
+#include "versioninfo.h"
 
 #include <locale.h>
 #include <stdlib.h>
 #include <string.h>
 
-#define PACKAGE "cinelerra"
-#define LOCALEDIR "/locale/"
+#include <locale.h>
+
+#define PACKAGE "cinecutie"
+#define LOCALEDIR "/usr/share/locale"
 
 
 enum
@@ -56,24 +59,6 @@
 
 #include "thread.h"
 
-void get_exe_path(char *result)
-{
-// Get executable path
-	pid_t pid = getpid();
-	char proc_path[BCTEXTLEN];
-	int len = 0;
-	result[0] = 0;
-	sprintf(proc_path, "/proc/%d/exe", pid);
-	if((len = readlink(proc_path, result, BCTEXTLEN)) >= 0)
-	{
-		result[len] = 0;
-//printf("Preferences::Preferences %d %s\n", __LINE__, result);
-		char *ptr = strrchr(result, '/');
-		if(ptr) *ptr = 0;
-	}
-
-}
-
 int main(int argc, char *argv[])
 {
 // handle command line arguments first
@@ -81,58 +66,22 @@
 	ArrayList<char*> filenames;
 	FileSystem fs;
 
-
 	int operation = DO_GUI;
 	int deamon_port = DEAMON_PORT;
 	char deamon_path[BCTEXTLEN];
 	char config_path[BCTEXTLEN];
 	char batch_path[BCTEXTLEN];
-	char locale_path[BCTEXTLEN];
-	char exe_path[BCTEXTLEN];
 	int nice_value = 20;
 	config_path[0] = 0;
 	batch_path[0] = 0;
 	deamon_path[0] = 0;
 	Garbage::garbage = new Garbage;
 	EDL::id_lock = new Mutex("EDL::id_lock");
-
-
-	get_exe_path(exe_path);
-	sprintf(locale_path, "%s%s", exe_path, LOCALEDIR);
-// detect an UTF-8 locale and try to use a non-Unicode locale instead
-// <---Beginning of dirty hack
-// This hack will be removed as soon as Cinelerra is UTF-8 compliant
-    char *s, *language;
-
-// Query user locale
-    if ((s = getenv("LC_ALL"))  || 
-		(s = getenv("LC_MESSAGES")) || 
-		(s = getenv("LC_CTYPE")) || 
-		(s = getenv ("LANG")))
-    {
-// Test if user locale is set to Unicode        
-        if (strstr(s, ".UTF-8"))
-        {
-// extract language  from language-charset@variant
-          language = strtok (s, ".@");
-// set language as the default locale
-          setenv("LANG", language, 1);
-        }
-    }
-// End of dirty hack --->
-
-	bindtextdomain (PACKAGE, locale_path);
+	bindtextdomain (PACKAGE, LOCALEDIR);
 	textdomain (PACKAGE);
 	setlocale (LC_MESSAGES, "");
 	setlocale (LC_CTYPE, "");
 
-
-
-
-
-
-
-
 	for(int i = 1; i < argc; i++)
 	{
 		if(!strcmp(argv[i], "-h"))
@@ -206,7 +155,7 @@
 		else
 		{
 			char *new_filename;
-			new_filename = new char[BCTEXTLEN];
+			new_filename = new char[1024];
 			strcpy(new_filename, argv[i]);
             fs.complete_path(new_filename);
 
@@ -224,8 +173,9 @@
 		operation == DO_BATCHRENDER)
 	fprintf(stderr, 
 		PROGRAM_NAME " " 
-		CINELERRA_VERSION " " 
-		"(C)2008 Adam Williams\n\n"
+		CINECUTIE_VERSION " " 
+		REPOMAINTXT
+		"\n\n"
 
 PROGRAM_NAME " is free software, covered by the GNU General Public License,\n"
 "and you are welcome to change it and/or distribute copies of it under\n"
@@ -300,28 +250,20 @@
 			mwindow.create_objects(1, 
 				!filenames.total,
 				config_path);
-//SET_TRACE
 
 // load the initial files on seperate tracks
 			if(filenames.total)
 			{
-//SET_TRACE
 				mwindow.gui->lock_window("main");
-//SET_TRACE
 				mwindow.load_filenames(&filenames, LOAD_REPLACE);
-//SET_TRACE
 				if(filenames.total == 1)
 					mwindow.gui->mainmenu->add_load(filenames.values[0]);
 				mwindow.gui->unlock_window();
-//SET_TRACE
 			}
 
 // run the program
-//SET_TRACE
 			mwindow.start();
-//SET_TRACE
 			mwindow.save_defaults();
-//SET_TRACE
 DISABLE_BUFFER
 			break;
 		}
