--- cinecutie/mainindexes.C	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/mainindexes.C	2010-02-19 22:19:11.968571001 +0100
@@ -67,14 +67,11 @@
 {
 	next_lock->lock("MainIndexes::add_next_asset");
 
-SET_TRACE
 // Test current asset
 	IndexFile indexfile(mwindow);
 
-SET_TRACE
 	int got_it = 0;
 
-SET_TRACE
 	if(!indexfile.open_index(asset))
 	{
 		asset->index_status = INDEX_READY;
@@ -82,12 +79,10 @@
 		got_it = 1;
 	}
 
-SET_TRACE
 	if(!got_it)
 	{
 		File *this_file = file;
 
-SET_TRACE
 		if(!file)
 		{
 			this_file = new File;
@@ -99,52 +94,35 @@
 				0);
 		}
 
-
-SET_TRACE
 		char index_filename[BCTEXTLEN];
 		char source_filename[BCTEXTLEN];
-SET_TRACE
 		IndexFile::get_index_filename(source_filename, 
 			mwindow->preferences->index_directory, 
 			index_filename, 
 			asset->path);
-SET_TRACE
 		if(!this_file->get_index(index_filename))
 		{
-SET_TRACE
 			if(!indexfile.open_index(asset))
 			{
-SET_TRACE
 				indexfile.close_index();
-SET_TRACE
 				asset->index_status = INDEX_READY;
 				got_it = 1;
 			}
-SET_TRACE
 		}
-SET_TRACE
 		if(!file) delete this_file;
-SET_TRACE
 	}
-SET_TRACE
 
 
 // Put copy of asset in stack, not the real thing.
 	if(!got_it)
 	{
 //printf("MainIndexes::add_next_asset 3\n");
-SET_TRACE
 		Asset *new_asset = new Asset;
-SET_TRACE
 		*new_asset = *asset;
 // If the asset existed and was overwritten, the status will be READY.
-//printf("MainIndexes::add_next_asset 1\n");
 		new_asset->index_status = INDEX_NOTTESTED;
-SET_TRACE
 		next_assets.append(new_asset);
-SET_TRACE
 	}
-SET_TRACE
 
 	next_lock->unlock();
 }
