--- cinecutie/menuaeffects.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/menuaeffects.h	2010-02-19 22:19:11.968571001 +0100
@@ -39,7 +39,7 @@
 class MenuAEffectThread : public MenuEffectThread
 {
 public:
-	MenuAEffectThread(MWindow *mwindow, MenuAEffects *menu_item);
+	MenuAEffectThread(MWindow *mwindow);
 	~MenuAEffectThread();
 
 	int get_recordable_tracks(Asset *asset);
