--- cinecutie/menueffects.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/menueffects.h	2010-02-19 22:19:11.968571001 +0100
@@ -67,11 +67,11 @@
 class MenuEffectThread : public Thread
 {
 public:
-	MenuEffectThread(MWindow *mwindow, MenuEffects *menu_item);
+	MenuEffectThread(MWindow *mwindow);
 	virtual ~MenuEffectThread();
 
 	void run();
-	int set_title(const char *text);  // set the effect to be run by a menuitem
+	int set_title(char *text);  // set the effect to be run by a menuitem
 	virtual int get_recordable_tracks(Asset *asset) { return 0; };
 	virtual int get_derived_attributes(Asset *asset, BC_Hash *defaults) { return 0; };
 	virtual int save_derived_attributes(Asset *asset, BC_Hash *defaults) { return 0; };
@@ -81,9 +81,8 @@
 	int test_existence(Asset *asset);
 
 	MWindow *mwindow;
-	MenuEffects *menu_item;
-	char title[BCTEXTLEN];
-	int realtime, load_mode;
+	char title[1024];
+	int dither, realtime, load_mode;
 	int strategy;
 };
 
@@ -117,7 +116,7 @@
 		Asset *asset);
 	virtual ~MenuEffectWindow();
 
-	void create_objects();
+	int create_objects();
 	int resize_event(int w, int h);
 
 	BC_Title *list_title;
@@ -181,7 +180,7 @@
 
 	static int calculate_w(BC_WindowBase *gui);
 	static int calculate_h(BC_WindowBase *gui);	
-	void create_objects();
+	int create_objects();
 
 	MenuEffectPromptOK *ok;
 	MenuEffectPromptCancel *cancel;
