--- cinecutie/module.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/module.C	2010-02-19 22:19:11.968571001 +0100
@@ -248,7 +248,7 @@
 	transition = track->get_current_transition(current_position,
 		direction,
 		0,
-		0);
+		0); // position is already nudged in amodule.C and vmodule.C before calling update_transition!
 
 // For situations where we had a transition but not anymore, 
 // keep the server open.
@@ -258,13 +258,18 @@
 
 // If the current transition differs from the previous transition, delete the
 // server.
-	if ((transition && 
-		transition_server && 
-		strcmp(transition->title, transition_server->plugin->title)))
+	if (transition && 
+		transition_server)
 	{
-		transition_server->close_plugin();
-		delete transition_server;
-		transition_server = 0;
+		if (strcmp(transition->title, transition_server->plugin->title))
+		{
+			transition_server->close_plugin();
+			delete transition_server;
+			transition_server = 0;
+		} else
+		{
+			transition_server->plugin = transition;
+		}
 	}
 
 	if(transition && !transition_server)
