--- cinecutie/mtimebar.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/mtimebar.C	2010-02-19 22:19:11.968571001 +0100
@@ -31,6 +31,7 @@
 #include "mtimebar.h"
 #include "mwindowgui.h"
 #include "mwindow.h"
+#include "patchbay.h"
 #include "preferences.h"
 #include "theme.h"
 #include "trackcanvas.h"
@@ -451,11 +452,13 @@
 	}
 
 // Que the CWindow
-	mwindow->cwindow->update(1, 0, 0);
+	mwindow->cwindow->update(1, 0, 0, 0, 1);
 	mwindow->gui->cursor->hide(0);
 	mwindow->gui->cursor->draw(1);
 	mwindow->gui->canvas->activate();
 	mwindow->gui->zoombar->update();
+	mwindow->gui->patchbay->update();
+	mwindow->update_plugin_guis();
 	update_highlights();
 	mwindow->gui->canvas->flash();
 }
