--- cinecutie/overlayframe.h	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/overlayframe.h	2010-02-19 22:19:11.978570999 +0100
@@ -106,7 +106,7 @@
 	void tabulate_bcubic_f(float* &coef_table, 
 		int* &coord_table,
 		float scale,
-		int start, 
+		float start, 
 		int pixels,
 		int total_pixels,
 		float coefficient);
@@ -117,7 +117,7 @@
 		float scale,
 		int pixel1,
 		int pixel2,
-		int start,
+		float start,
 		int total_pixels);
 
 	void tabulate_bcubic_i(int* &coef_table, 
@@ -134,7 +134,7 @@
 		float scale,
 		int pixel1,
 		int pixel2,
-		int start,
+		float start,
 		int total_pixels);
 	void tabulate_reduction(bilinear_table_t* &table,
 		float scale,
@@ -143,7 +143,7 @@
 		int in_total);
 	void tabulate_enlarge(bilinear_table_t* &table,
 		float scale,
-		int pixel1,
+		float pixel1,
 		int out_total,
 		int in_total);
 	void dump_bilinear(bilinear_table_t *table, int total);
@@ -170,8 +170,8 @@
 	VFrame *scale_input;
 	float w_scale;
 	float h_scale;
-	int in_x1_int;
-	int in_y1_int;
+	float in_x1_float;
+	float in_y1_float;
 	int out_w_int;
 	int out_h_int;
 	int interpolation_type;
@@ -290,12 +290,11 @@
 	~ScaleTranslateUnit();
 
 	void process_package(LoadPackage *package);
-	void scale_array(int* &table, 
+	void scale_array_f(int* &table, 
 		int out_x1, 
 		int out_x2,
-		int in_x1,
-		int in_x2,
-		int is_x);
+		float in_x1,
+		float in_x2);
 	
 	OverlayFrame *overlay;
 	ScaleTranslateEngine *scale_translate;
@@ -317,10 +316,10 @@
 // Arguments
 	VFrame *output;
 	VFrame *input;
-	int in_x1;
-	int in_y1;
-	int in_x2;
-	int in_y2;
+	float in_x1;
+	float in_y1;
+	float in_x2;
+	float in_y2;
 	int out_x1;
 	int out_y1;
 	int out_x2;
