--- cinecutie/overlayframe.inc	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/overlayframe.inc	2010-02-19 22:19:11.978570999 +0100
@@ -40,7 +40,7 @@
 #define NEAREST_NEIGHBOR 0        // Nearest for enlargement and reduction
 #define CUBIC_CUBIC      1        // Bicubic enlargement and reduction
 #define CUBIC_LINEAR     2        // Bicubic enlargement and linear reduction
-#define LINEAR_LINEAR    3        // Not explicitely called but provided for user
+#define LINEAR_LINEAR    3        // Not explicitely called but used for user
 
 class OverlayFrame;
 
