--- cinecutie/patchbay.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/patchbay.h	2010-02-19 22:19:11.978570999 +0100
@@ -42,7 +42,7 @@
 	~PatchBay();
 
 	int delete_all_patches();
-	void create_objects();
+	int create_objects();
 	void resize_event();
 	int button_press_event();
 	int cursor_motion_event();
@@ -70,7 +70,7 @@
 // =========================================== drawing
 
 	int resize_event(int top, int bottom);
-
+	Track *is_over_track();	// called from trackcanvas!
 
 	MWindow *mwindow;
 	MWindowGUI *gui;
