--- cinecutie/pluginset.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/pluginset.h	2010-02-19 22:19:11.988570758 +0100
@@ -55,13 +55,13 @@
 	void copy_keyframes(int64_t start, 
 		int64_t end, 
 		FileXML *file, 
-		int use_default,
-		int active_only);
-	void paste_keyframes(int64_t start, 
+		int default_only,
+		int autos_only);
+	static void paste_keyframes(int64_t start, 
 		int64_t length, 
 		FileXML *file, 
-		int use_default,
-		int active_only);
+		int default_only,
+		Track *track);
 // Return the nearest boundary of any kind in the plugin edits
 	int64_t plugin_change_duration(int64_t input_position, 
 		int64_t input_length, 
@@ -79,7 +79,7 @@
 	int optimize();
 
 // Insert a new plugin
-	Plugin* insert_plugin(const char *title, 
+	Plugin* insert_plugin(char *title, 
 		int64_t unit_position, 
 		int64_t unit_length,
 		int plugin_type,
