--- cinecutie/pluginpopup.C	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/pluginpopup.C	2010-02-19 22:19:11.988570758 +0100
@@ -26,8 +26,6 @@
 #include "plugin.h"
 #include "plugindialog.h"
 #include "pluginpopup.h"
-#include "presets.h"
-#include "presetsgui.h"
 #include "track.h"
 
 
@@ -41,9 +39,6 @@
 {
 	this->mwindow = mwindow;
 	this->gui = gui;
-#if 0
-	thread = new PresetsThread(mwindow);
-#endif
 }
 
 PluginPopup::~PluginPopup()
@@ -54,26 +49,21 @@
 {
 	add_item(change = new PluginPopupChange(mwindow, this));
 	add_item(detach = new PluginPopupDetach(mwindow, this));
+//	add_item(in = new PluginPopupIn(mwindow, this));
+//	add_item(out = new PluginPopupOut(mwindow, this));
+	add_item(show = new PluginPopupShow(mwindow, this));
+	add_item(on = new PluginPopupOn(mwindow, this));
 	add_item(new PluginPopupUp(mwindow, this));
 	add_item(new PluginPopupDown(mwindow, this));
-	add_item(on = new PluginPopupOn(mwindow, this));
 }
 
 int PluginPopup::update(Plugin *plugin)
 {
-	if(show) remove_item(show);
-	if(presets) remove_item(presets);
-	show = 0;
-	presets = 0;
-
-	if(plugin->plugin_type == PLUGIN_STANDALONE)
-	{
-		add_item(show = new PluginPopupShow(mwindow, this));
-		add_item(presets = new PluginPresets(mwindow, this));
-		show->set_checked(plugin->show);
-	}
-
+//printf("PluginPopup::update %p\n", plugin);
 	on->set_checked(plugin->on);
+//	in->set_checked(plugin->in);
+//	out->set_checked(plugin->out);
+	show->set_checked(plugin->show);
 	this->plugin = plugin;
 	return 0;
 }
@@ -86,7 +76,8 @@
 
 
 
-PluginPopupChange::PluginPopupChange(MWindow *mwindow, PluginPopup *popup)
+PluginPopupChange::PluginPopupChange(MWindow *mwindow, PluginPopup
+*popup)
  : BC_MenuItem(_("Change..."))
 {
 	this->mwindow = mwindow;
@@ -103,9 +94,7 @@
 {
 	dialog_thread->start_window(popup->plugin->track,
 		popup->plugin,
-		PROGRAM_NAME ": Change Effect",
-		0,
-		popup->plugin->track->data_type);
+		PROGRAM_NAME ": Change Effect");
 }
 
 
@@ -128,12 +117,10 @@
 
 int PluginPopupDetach::handle_event()
 {
-	mwindow->undo->update_undo_before();
 	mwindow->hide_plugin(popup->plugin, 1);
-	mwindow->hide_keyframe_gui(popup->plugin);
 	popup->plugin->track->detach_effect(popup->plugin);
 	mwindow->save_backup();
-	mwindow->undo->update_undo_after(_("detach effect"), LOAD_ALL);
+	mwindow->undo->update_undo(_("detach effect"), LOAD_ALL);
 	mwindow->gui->update(0,
 		1,
 		0,
@@ -210,7 +197,6 @@
 int PluginPopupShow::handle_event()
 {
 	mwindow->show_plugin(popup->plugin);
-	mwindow->gui->update(0, 1, 0, 0, 0, 0, 0);
 	return 1;
 }
 
@@ -266,21 +252,3 @@
 	return 1;
 }
 
-
-
-PluginPresets::PluginPresets(MWindow *mwindow, PluginPopup *popup)
- : BC_MenuItem(_("Presets..."))
-{
-	this->mwindow = mwindow;
-	this->popup = popup;
-}
-
-int PluginPresets::handle_event()
-{
-	mwindow->show_keyframe_gui(popup->plugin);
-#if 0
-	popup->thread->start_window(popup->plugin);
-#endif
-	return 1;
-}
-
