--- cinecutie/preferences.h	2008-08-06 20:45:37.000000000 +0200
+++ /home/user/cinecutie/cinecutie/preferences.h	2010-02-19 22:19:11.988570758 +0100
@@ -60,7 +60,7 @@
 	void sort_nodes();
 	void edit_node(int number, char *new_text, int port, int enabled);
 	int get_enabled_nodes();
-	const char* get_node_hostname(int number);
+	char* get_node_hostname(int number);
 	int get_node_port(int number);
 // Copy frame rates.  Always used where the argument is the renderfarm and this is
 // the master preferences.  This way, the value for master node is properly 
@@ -125,9 +125,9 @@
 // Tip of the day
 	int use_tipwindow;
 
-
 // ====================================== Plugin Set ==============================
-	char plugin_dir[BCTEXTLEN];
+	char global_plugin_dir[BCTEXTLEN];
+	char local_plugin_dir[BCTEXTLEN];
 
 // Required when updating renderfarm rates
 	Mutex *preferences_lock;
