--- cinecutie/recordgui.C	2008-10-11 17:46:25.000000000 +0200
+++ /home/user/cinecutie/cinecutie/recordgui.C	2010-02-19 22:19:11.988570758 +0100
@@ -75,7 +75,6 @@
 RecordGUI::~RecordGUI()
 {
 TRACE("RecordGUI::~RecordGUI 1");
-	lock_window("RecordGUI::~RecordGUI");
 	delete status_thread;
 	delete batch_source;
 	delete batch_mode;
@@ -84,12 +83,11 @@
 	delete batch_start;
 	delete batch_duration;
 	delete load_mode;
-	unlock_window();
 TRACE("RecordGUI::~RecordGUI 2");
 }
 
 
-const char* RecordGUI::batch_titles[] = 
+char* RecordGUI::batch_titles[] = 
 {
 	N_("On"),
 	N_("Path"),
@@ -132,12 +130,11 @@
 }
 
 
-void RecordGUI::create_objects()
+int RecordGUI::create_objects()
 {
 	char string[BCTEXTLEN];
 	flash_color = RED;
 
-	lock_window("RecordGUI::create_objects");
 	status_thread = new RecordStatusThread(mwindow, this);
 	status_thread->start();
 	set_icon(mwindow->theme->get_image("record_icon"));
@@ -404,7 +401,7 @@
 	if(monitor_audio) y += monitor_audio->get_h();
 
 	int bottom_margin = MAX(BC_OKButton::calculate_h(), 
-		LoadMode::calculate_h(this, mwindow->theme)) + 5;
+		LoadMode::calculate_h(this)) + 5;
 
 
 	add_subwindow(batch_list = new RecordGUIBatches(record, 
@@ -416,10 +413,9 @@
 	y += batch_list->get_h() + 5;
 
 // Controls
-	int loadmode_w = LoadMode::calculate_w(this, mwindow->theme, 1);
 	load_mode = new LoadMode(mwindow,
 		this, 
-		get_w() / 2 - loadmode_w / 2, 
+		get_w() / 2 - mwindow->theme->loadmode_w / 2, 
 		y, 
 		&record->load_mode, 
 		1);
@@ -434,7 +430,7 @@
 
 	startover_thread = new RecordStartoverThread(record, this);
 
-	unlock_window();
+	return 0;
 }
 
 void RecordGUI::flash_batch()
@@ -566,6 +562,7 @@
 
 	int new_h = mwindow->session->rwindow_h - bottom_margin - batch_list->get_y();
 	if(new_h < 10) new_h = 10;
+printf("RecordGUI::resize_event 1 %d\n", mwindow->session->rwindow_h - bottom_margin - batch_list->get_y());
 	batch_list->reposition_window(batch_list->get_x(), 
 		batch_list->get_y(),
 		mwindow->session->rwindow_w - 20,
