--- cinecutie/recordmonitor.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/recordmonitor.h	2010-02-19 22:19:11.988570758 +0100
@@ -22,7 +22,9 @@
 #ifndef RECORDMONITOR_H
 #define RECORDMONITOR_H
 
+#ifdef HAVE_FIREWIRE
 #include "avc1394transport.h"
+#endif
 #include "canvas.h"
 #include "channelpicker.inc"
 #include "condition.inc"
@@ -71,7 +73,7 @@
 	void run();
 
 	int close_threads();   // Stop all the child threads on exit
-	void create_objects();
+	int create_objects();
 	int fix_size(int &w, int &h, int width_given, float aspect_ratio);
 	float get_scale(int w);
 	int get_mbuttons_height();
@@ -92,7 +94,7 @@
 		int min_w);
 	~RecordMonitorGUI();
 
-	void create_objects();
+	int create_objects();
 	int cursor_leave_event();
 	int cursor_enter_event();
 	int button_release_event();
@@ -101,8 +103,10 @@
 	MeterPanel *meters;
 	Canvas *canvas;
 //	RecordTransport *record_transport;
+#ifdef HAVE_FIREWIRE
 	AVC1394Transport *avc1394_transport;
 	AVC1394TransportThread *avc1394transport_thread;
+#endif
 	RecordChannelPicker *channel_picker;
 	ReverseInterlace *reverse_interlace;
 	int cursor_x_origin, cursor_y_origin;
@@ -136,9 +140,11 @@
 	BC_Bitmap *bitmap;
 	RecordMonitor *thread;
 	Record *record;
+#ifdef HAVE_FIREWIRE
 	AVC1394Control *avc;
 	BC_Title *avc1394transport_title;
 	BC_Title *avc1394transport_timecode;
+#endif
 };
 
 
