--- cinecutie/recordwindow.C	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/recordwindow.C	2010-02-19 22:19:11.988570758 +0100
@@ -56,7 +56,7 @@
 
 
 
-void RecordWindow::create_objects()
+int RecordWindow::create_objects()
 {
 //printf("RecordWindow::create_objects 1\n");
 	add_subwindow(new BC_Title(5, 5, _("Select a file to record to:")));
@@ -92,6 +92,8 @@
 	add_subwindow(new BC_CancelButton(this));
 //printf("RecordWindow::create_objects 1\n");
 	show_window();
+//printf("RecordWindow::create_objects 2\n");
+	return 0;
 }
 
 
