--- cinecutie/render.inc	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/render.inc	2010-02-19 22:19:11.998570756 +0100
@@ -26,7 +26,7 @@
 class Render;
 class RenderItem;
 class RenderJob;
-
+class RenderWindow;
 
 // Internal rendering strategies
 
@@ -37,5 +37,9 @@
 #define BRENDER_FARM            4
 
 
+#define RANGE_BACKCOMPAT	0
+#define RANGE_PROJECT		1
+#define RANGE_SELECTION		2
+#define RANGE_INOUT		3
 
 #endif
