--- cinecutie/resizetrackthread.C	2008-10-13 20:50:43.000000000 +0200
+++ /home/user/cinecutie/cinecutie/resizetrackthread.C	2010-02-19 22:19:11.998570756 +0100
@@ -130,7 +130,6 @@
 {
 	int x = 10, y = 10;
 
-	lock_window("ResizeTrackWindow::create_objects");
 	add_subwindow(new BC_Title(x, y, _("Size:")));
 	x += 50;
 	add_subwindow(w = new ResizeTrackWidth(this, 
@@ -176,7 +175,6 @@
 
 	show_window();
 	flush();
-	unlock_window();
 }
 
 void ResizeTrackWindow::update(int changed_scale, 
