--- cinecutie/resourcethread.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/resourcethread.C	2010-02-19 22:19:11.998570756 +0100
@@ -349,13 +349,13 @@
 	}
 	else
 	{
+
 		File *source = mwindow->video_cache->check_out(item->asset,
 			mwindow->edl);
 		if(!source) 
 		{
 			return;
 		}
-
 		source->set_layer(item->layer);
 		source->set_video_position(item->position, 
 			item->frame_rate);
