--- cinecutie/sharedlocation.C	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/sharedlocation.C	2010-02-19 22:19:11.998570756 +0100
@@ -21,7 +21,6 @@
 
 #include "edl.h"
 #include "filexml.h"
-#include "language.h"
 #include "plugin.h"
 #include "sharedlocation.h"
 #include "track.h"
@@ -33,6 +32,10 @@
 
 #include <string.h>
 
+#include <libintl.h>
+#define _(String) gettext(String)
+#define gettext_noop(String) String
+#define N_(String) gettext_noop (String)
 
 
 SharedLocation::SharedLocation()
@@ -78,7 +81,6 @@
 
 int SharedLocation::operator==(const SharedLocation &that)
 {
-//printf("SharedLocation::operator== called\n");
 	if(
 		module == that.module &&
 		plugin == that.plugin
@@ -87,32 +89,13 @@
 	return 0;
 }
 
-int SharedLocation::equivalent(SharedLocation *that)
-{
-	if(
-		module == that->module &&
-		plugin == that->plugin
-	) return 1;
-	else
-	return 0;
-	
-}
-
 SharedLocation& SharedLocation::operator=(const SharedLocation &that)
 {
-//printf("SharedLocation::operator= called\n");
 	this->plugin = that.plugin;
 	this->module = that.module;
 	return *this;
 }
 
-void SharedLocation::copy_from(SharedLocation *that)
-{
-	this->plugin = that->plugin;
-	this->module = that->module;
-}
-
-
 void SharedLocation::calculate_title(char *string, 
 	EDL *edl, 
 	double position, 
