--- cinecutie/statusbar.C	2008-09-02 13:38:13.000000000 +0200
+++ /home/user/cinecutie/cinecutie/statusbar.C	2010-02-19 22:19:11.998570756 +0100
@@ -48,7 +48,7 @@
 
 
 
-void StatusBar::create_objects()
+int StatusBar::create_objects()
 {
 //printf("StatusBar::create_objects 1\n");
 	int x = 10, y = 5;
@@ -60,8 +60,7 @@
 		MEDIUMFONT,
 		mwindow->theme->message_normal));
 	x = get_w() - 290;
-// printf("StatusBar::create_objects %d: 0x%08x\n", 
-// __LINE__, mwindow->theme->message_normal);
+//printf("StatusBar::create_objects 1\n");
 	add_subwindow(main_progress = 
 		new BC_ProgressBar(mwindow->theme->mstatus_progress_x, 
 			mwindow->theme->mstatus_progress_y, 
@@ -76,6 +75,8 @@
 //printf("StatusBar::create_objects 1\n");
 	default_message();
 	flash();
+//printf("StatusBar::create_objects 2\n");
+	return 0;
 }
 
 void StatusBar::resize_event()
@@ -108,7 +109,7 @@
 void StatusBar::default_message()
 {
 	status_text->set_color(mwindow->theme->message_normal);
-	status_text->update(_("Welcome to Cinelerra."));
+	status_text->update(_("Welcome to Cinecutie."));
 }
 
 
