--- cinecutie/theme.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/theme.h	2010-02-19 22:19:12.008571675 +0100
@@ -27,7 +27,6 @@
 #include "bctheme.h"
 #include "cwindowgui.inc"
 #include "guicast.h"
-#include "keyframegui.inc"
 #include "levelwindowgui.inc"
 #include "mbuttons.inc"
 #include "mwindow.inc"
@@ -36,7 +35,6 @@
 #include "overlayframe.inc"
 #include "patchbay.inc"
 #include "preferencesthread.inc"
-#include "presetsgui.inc"
 #include "recordgui.inc"
 #include "recordmonitor.inc"
 #include "resourcepixmap.inc"
@@ -83,8 +81,6 @@
 		int w, 
 		int h);
 	virtual void get_plugindialog_sizes();
-//	virtual void get_presetdialog_sizes(PresetsWindow *gui);
-	virtual void get_keyframedialog_sizes(KeyFrameWindow *gui);
 	virtual void get_menueffect_sizes(int use_list);
 	virtual void draw_rwindow_bg(RecordGUI *gui);
 	virtual void draw_rmonitor_bg(RecordMonitorGUI *gui);
@@ -118,7 +114,7 @@
 	ArrayList<BC_ListBoxItem*> frame_sizes;
 	ArrayList<BC_ListBoxItem*> sample_rates;
 	ArrayList<BC_ListBoxItem*> zoom_values;
-	const char *theme_title;
+	char *theme_title;
 
 // Tools for building widgets
 	void overlay(VFrame *dst, VFrame *src, int in_x1 = -1, int in_x2 = -1);
@@ -200,8 +196,6 @@
 	int patchbay_x, patchbay_y, patchbay_w, patchbay_h;
 // pixels between toggles and buttons in edit panel
 	int toggle_margin;
-// Distance to left and right of automation point for control points
-	int control_pixels;
 
 	int plugindialog_new_x, plugindialog_new_y, plugindialog_new_w, plugindialog_new_h;
 	int plugindialog_shared_x, plugindialog_shared_y, plugindialog_shared_w, plugindialog_shared_h;
@@ -210,22 +204,6 @@
 	int plugindialog_sharedattach_x, plugindialog_sharedattach_y;
 	int plugindialog_moduleattach_x, plugindialog_moduleattach_y;
 
-
-
-
-
-	int presets_list_x, presets_list_y, presets_list_w, presets_list_h;
-	int presets_text_x, presets_text_y, presets_text_w;
-	int presets_delete_x, presets_delete_y;
-	int presets_save_x, presets_save_y;
-	int presets_apply_x, presets_apply_y;   
-
-
-	int keyframe_list_x, keyframe_list_y, keyframe_list_w, keyframe_list_h;
-	int keyframe_text_x, keyframe_text_y, keyframe_text_w, keyframe_text_h;
-	int keyframe_value_x, keyframe_value_y, keyframe_value_w, keyframe_value_h;
-	int keyframe_all_x, keyframe_all_y;
-
 	int menueffect_list_x, menueffect_list_y, menueffect_list_w, menueffect_list_h;
 	int menueffect_file_x, menueffect_file_y;
 	int menueffect_tools_x, menueffect_tools_y;
@@ -257,6 +235,7 @@
 	int setformat_channels_x, setformat_channels_y, setformat_channels_w, setformat_channels_h;
 	int title_h;
 	int title_font, title_color;
+	int edit_font_color;
 	int vcanvas_x, vcanvas_y, vcanvas_w, vcanvas_h;
 	int vedit_x, vedit_y;
 	int vmeter_x, vmeter_y, vmeter_h;
@@ -266,8 +245,7 @@
 	int vtime_x, vtime_y, vtime_w;
 	int vtransport_x, vtransport_y;
 	int vzoom_x, vzoom_y;
-	int window_border;
-	int widget_border;
+
 
 
 // Bitmaps
