--- cinecutie/threadindexer.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/threadindexer.C	2010-02-19 22:19:12.008571675 +0100
@@ -143,15 +143,13 @@
 			}
 			else
 			{
-				if(current_asset->index_status == INDEX_NOTTESTED) 
-					current_asset->index_status = INDEX_READY;   // index has been tested
+				if(current_asset->index_status == 1) current_asset->index_status = 0;   // index has been tested
 				indexfile->close_index();
 			}
 		}
 	}
 
-// progress box is only created when an index is built
-	if(progress)     
+	if(progress)     // progress box is only createdd when an index is built
 	{	
 		progress->stop_progress();
 		delete progress;
