--- cinecutie/threadfork.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/threadfork.C	2010-02-19 22:19:12.008571675 +0100
@@ -74,7 +74,7 @@
 	threadfork->run();
 }
 
-void ThreadFork::start_command(const char *command_line, int pipe_stdin)
+void ThreadFork::start_command(char *command_line, int pipe_stdin)
 {
 	this->command_line = command_line;
 	this->pipe_stdin = pipe_stdin;
