--- cinecutie/timebar.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/timebar.C	2010-02-19 22:19:12.008571675 +0100
@@ -29,6 +29,7 @@
 #include "filexml.h"
 #include "fonts.h"
 #include "labels.h"
+#include "labeledit.h"
 #include "localsession.h"
 #include "maincursor.h"
 #include "mainundo.h"
@@ -68,6 +69,7 @@
 	this->gui = 0;
 	this->pixel = pixel;
 	this->position = position;
+	this->label = 0;
 }
 
 LabelGUI::~LabelGUI()
@@ -91,6 +93,18 @@
 	reposition_window(translate_pixel(mwindow, pixel), BC_Toggle::get_y());
 }
 
+int LabelGUI::button_press_event()
+{
+	if (this->is_event_win() && get_buttonpress() == 3) {
+		if (label)
+			timebar->label_edit->edit_label(label);
+	} else {
+		BC_Toggle::button_press_event();
+	}
+	if (label)
+		set_tooltip(this->label->textstr);
+}
+
 int LabelGUI::handle_event()
 {
 	timebar->select_label(position);
@@ -180,22 +194,25 @@
 //printf("TimeBar::TimeBar %d %d %d %d\n", x, y, w, h);
 	this->gui = gui;
 	this->mwindow = mwindow;
+	label_edit = new LabelEdit(mwindow, mwindow->awindow, 0);
 }
 
 TimeBar::~TimeBar()
 {
 	if(in_point) delete in_point;
 	if(out_point) delete out_point;
+	if(label_edit) delete label_edit;
 	labels.remove_all_objects();
 	presentations.remove_all_objects();
 }
 
-void TimeBar::create_objects()
+int TimeBar::create_objects()
 {
 	in_point = 0;
 	out_point = 0;
 	current_operation = TIMEBAR_NONE;
 	update();
+	return 0;
 }
 
 
@@ -232,6 +249,8 @@
 							LabelGUI::get_y(mwindow, this), 
 							current->position));
 					new_label->set_cursor(ARROW_CURSOR);
+					new_label->set_tooltip(current->textstr);
+					new_label->label = current;
 					labels.append(new_label);
 				}
 				else
@@ -249,6 +268,8 @@
 					}
 
 					labels.values[output]->position = current->position;
+					labels.values[output]->set_tooltip(current->textstr);
+					labels.values[output]->label = current;
 				}
 
 				if(edl->local_session->get_selectionstart(1) <= current->position &&
@@ -839,6 +860,7 @@
 	switch(current_operation)
 	{
 		case TIMEBAR_DRAG:
+			mwindow->undo->update_undo(_("select"), LOAD_SESSION, 0, 0);
 			mwindow->gui->canvas->stop_dragscroll();
 			current_operation = TIMEBAR_NONE;
 			result = 1;
@@ -922,6 +944,7 @@
 	mwindow->gui->canvas->flash();
 	mwindow->gui->canvas->activate();
 	mwindow->gui->zoombar->update();
+	mwindow->undo->update_undo(_("select"), LOAD_SESSION, 0, 0);
 	update_highlights();
 	return 0;
 }
