--- cinecutie/tracks.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/tracks.h	2010-02-19 22:19:12.008571675 +0100
@@ -50,7 +50,8 @@
 		Track *track,
 		double position,
 		int edit_labels,
-		int edit_plugins);
+		int edit_plugins,
+		int behaviour);
 	void move_effect(Plugin *plugin,
 		PluginSet *plugin_set,
 		Track *track, 
@@ -64,7 +65,8 @@
 	void get_automation_extents(float *min, 
 		float *max,
 		double start,
-		double end);
+		double end,
+		int autogrouptype);
 
 	void equivalent_output(Tracks *tracks, double *result);
 
@@ -75,10 +77,6 @@
 	void paste_audio_transition(PluginServer *server);
 	void paste_video_transition(PluginServer *server, int first_track = 0);
 
-// Only tests effects
-	int plugin_exists(Plugin *plugin);
-	int track_exists(Track *track);
-
 	void paste_transition(PluginServer *server, Edit *dest_edit);
 // Return the numbers of tracks with the play patch enabled
 	int playable_audio_tracks();
@@ -90,7 +88,9 @@
 	int total_video_tracks();
 // return the longest track in all the tracks in seconds
  	double total_length();
+ 	double total_audio_length();
  	double total_video_length();
+	double total_length_framealigned(double fps);
 // Update y pixels after a zoom
 	void update_y_pixels(Theme *theme);
 // Total number of tracks where the following toggles are selected
@@ -158,7 +158,7 @@
 		double end, 
 		int all, 
 		FileXML *file, 
-		const char *output_path = "");
+		char *output_path = "");
 
 
 
@@ -169,9 +169,8 @@
 	int clear(double start, double end, int clear_plugins);
 	void clear_automation(double selectionstart, 
 		double selectionend);
-	void set_automation_mode(double selectionstart, 
-		double selectionend,
-		int mode);
+	void straighten_automation(double selectionstart, 
+		double selectionend);
 	int clear_default_keyframe();
 	int clear_handle(double start, 
 		double end,
@@ -183,12 +182,12 @@
 		FileXML *file,
 		int default_only,
 		int autos_only);
-//	int copy_default_keyframe(FileXML *file);
+	int copy_default_keyframe(FileXML *file);
+	void loaded_lengths_to_tracklengths(int includerecordtracks);
 	void paste_automation(double selectionstart, 
 		FileXML *xml,
-		int default_only,
-		int active_only);
-//	int paste_default_keyframe(FileXML *file);
+		int default_only);
+	int paste_default_keyframe(FileXML *file);
 	int paste(int64_t start, int64_t end);
 // all units are samples by default
 	int paste_output(int64_t startproject, 
@@ -227,11 +226,6 @@
 // Move edit boundaries and automation during a framerate change
 	int scale_time(float rate_scale, int ignore_record, int scale_edits, int scale_autos, int64_t start, int64_t end);
 
-	void clear_transitions(double start, double end);
-	void set_transition_length(double start, double end, double length);
-	void set_transition_length(Transition *transition, double length);
-	void paste_transitions(double start, double end, int track_type, char* title);
-
 // ================================== accounting
 
 	int handles, titles;     // show handles or titles
