--- cinecutie/tracks.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/tracks.C	2010-02-19 22:19:12.008571675 +0100
@@ -22,6 +22,7 @@
 #include "atrack.h"
 #include "automation.h"
 #include "cursor.h"
+#include "clip.h"
 #include "bchash.h"
 #include "edit.h"
 #include "edits.h"
@@ -137,7 +138,8 @@
 void Tracks::get_automation_extents(float *min, 
 	float *max,
 	double start,
-	double end)
+	double end,
+	int autogrouptype)
 {
 	*min = 0;
 	*max = 0;
@@ -149,8 +151,9 @@
 			current->automation->get_extents(min,
 				max,
 				&coords_undefined,
-				current->to_units(start, 1),
-				current->to_units(end, 1));
+				current->to_units(start, 0),
+				current->to_units(end, 1),
+				autogrouptype);
 		}
 	}
 }
@@ -461,6 +464,17 @@
 	return total; 
 }
 
+double Tracks::total_audio_length() 
+{
+	double total = 0;
+	for(Track *current = first; current; current = NEXT)
+	{
+		if(current->data_type == TRACK_AUDIO &&
+			current->get_length() > total) total = current->get_length();
+	}
+	return total; 
+}
+
 double Tracks::total_video_length() 
 {
 	double total = 0;
@@ -472,6 +486,19 @@
 	return total; 
 }
 
+double Tracks::total_length_framealigned(double fps)
+{
+	if (total_audio_tracks() && total_video_tracks())
+		return MIN(floor(total_audio_length() * fps), floor(total_video_length() * fps)) / fps;
+
+	if (total_audio_tracks())
+		return floor(total_audio_length() * fps) / fps;
+
+	if (total_video_tracks())
+		return floor(total_video_length() * fps) / fps;
+
+	return 0;
+}
 
 void Tracks::translate_camera(float offset_x, float offset_y)
 {
@@ -641,25 +668,3 @@
 	}
 	return result;
 }
-
-int Tracks::plugin_exists(Plugin *plugin)
-{
-	for(Track *track = first; track; track = track->next)
-	{
-		if(track->plugin_exists(plugin)) return 1;
-	}
-	return 0;
-}
-
-int Tracks::track_exists(Track *track)
-{
-	for(Track *current = first; current; current = NEXT)
-	{
-		if(current == track) return 1;
-	}
-	return 0;
-}
-
-
-
-
