--- cinecutie/transportque.h	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/transportque.h	2010-02-19 22:19:12.018571035 +0100
@@ -44,7 +44,11 @@
 	void set_playback_range(EDL *edl = 0, int use_inout = 0);
 
 // Adjust playback range with in/out points for rendering
-	void adjust_playback_range();
+	void playback_range_adjust_inout();
+// Set playback range to in/out points for rendering
+	void playback_range_inout();
+// Set playback range to whole project for rendering
+	void playback_range_project();
 
 	int single_frame();
 	EDL* get_edl();
