--- cinecutie/transportque.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/transportque.C	2010-02-19 22:19:12.018571035 +0100
@@ -169,8 +169,7 @@
 			else
 				end_position = edl->local_session->get_selectionend(1);
 // this prevents a crash if start position is after the loop when playing forwards
- 		    if (edl->local_session->loop_playback && 
-				start_position > edl->local_session->loop_end)  
+ 		    if (edl->local_session->loop_playback && start_position > edl->local_session->loop_end)  
  			{
 				    start_position = edl->local_session->loop_start;
 			}
@@ -184,11 +183,10 @@
 				start_position = 0;
 			else
 				start_position = edl->local_session->get_selectionstart(1);
-
 // this prevents a crash if start position is before the loop when playing backwards
-			if (edl->local_session->loop_playback && 
-				end_position <= edl->local_session->loop_start)
+			if (edl->local_session->loop_playback && start_position <= edl->local_session->loop_start)
 			{
+					start_position = edl->local_session->loop_end;
 					end_position = edl->local_session->loop_end;
 			}
 			break;
@@ -231,26 +229,33 @@
 
 }
 
-void TransportCommand::adjust_playback_range()
+void TransportCommand::playback_range_adjust_inout()
 {
-
-
 	if(edl->local_session->inpoint_valid() ||
 		edl->local_session->outpoint_valid())
 	{
-		if(edl->local_session->inpoint_valid())
-			start_position = edl->local_session->get_inpoint();
-		else
-			start_position = 0;
-
-		if(edl->local_session->outpoint_valid())
-			end_position = edl->local_session->get_outpoint();
-		else
-			end_position = edl->tracks->total_playable_length();
+		playback_range_inout();
 	}
 }
 
+void TransportCommand::playback_range_inout()
+{
+	if(edl->local_session->inpoint_valid())
+		start_position = edl->local_session->get_inpoint();
+	else
+		start_position = 0;
 
+	if(edl->local_session->outpoint_valid())
+		end_position = edl->local_session->get_outpoint();
+	else
+		end_position = edl->tracks->total_playable_length();
+}
+
+void TransportCommand::playback_range_project()
+{
+	start_position = 0;
+	end_position = edl->tracks->total_playable_length();
+}
 
 
 
