--- cinecutie/transition.h	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/transition.h	2010-02-19 22:19:12.018571035 +0100
@@ -58,7 +58,7 @@
 class Transition : public Plugin
 {
 public:
-	Transition(EDL *edl, Edit *edit, const char *title, long unit_length);
+	Transition(EDL *edl, Edit *edit, char *title, long unit_length);
 
 	Edit *edit;
 
@@ -92,7 +92,7 @@
 	int update_display();
 // Update edit after attaching
 	int update_edit(int is_loading);
-	const char* default_title();
+	char* default_title();
 	void dump();
 
 private:
