--- cinecutie/vautomation.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/vautomation.C	2010-02-19 22:19:12.018571035 +0100
@@ -45,7 +45,7 @@
 }
 
 
-void VAutomation::create_objects()
+int VAutomation::create_objects()
 {
 	Automation::create_objects();
 
@@ -79,6 +79,14 @@
 // 	autos[AUTOMATION_NUDGE] = new FloatAutos(edl, track, 0.0);
 // 	autos[AUTOMATION_NUDGE]->create_objects();
 
+	for(int i = 0; i < AUTOMATION_TOTAL; i++)
+		if (autos[i]) 
+		{
+			autos[i]->autoidx = i;
+			autos[i]->autogrouptype = autogrouptype(i, autos[i]->track);
+		}
+
+	return 0;
 }
 
 int VAutomation::direct_copy_possible(int64_t start, int direction)
@@ -91,7 +99,7 @@
 // Automation is constant
 	double constant;
 	if(((FloatAutos*)autos[AUTOMATION_FADE])->automation_is_constant(
-		start, end, direction, constant))
+		start, 1, direction, constant))
 	{
 		if(!EQUIV(constant, 100))
 			return 0;
