--- cinecutie/vdevicev4l2.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/vdevicev4l2.C	2010-02-19 22:19:12.018571035 +0100
@@ -24,7 +24,6 @@
 #undef _LARGEFILE64_SOURCE
 
 #include "assets.h"
-#include "bcsignals.h"
 #include "channel.h"
 #include "chantables.h"
 #include "clip.h"
@@ -310,6 +309,7 @@
 				perror("VDeviceV4L2Thread::run VIDIOC_G_PARM");
 		}
 
+
 // Set up data format
 		struct v4l2_format v4l2_params;
 		v4l2_params.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
@@ -383,14 +383,12 @@
 			}
 		}
 
-
 		for(int i = 0; i < picture->controls.total; i++)
 		{
 			struct v4l2_control ctrl_arg;
 			struct v4l2_queryctrl arg;
 			PictureItem *item = picture->controls.values[i];
 			arg.id = item->device_id;
-
 			if(!ioctl(input_fd, VIDIOC_QUERYCTRL, &arg))
 			{
 				ctrl_arg.id = item->device_id;
@@ -427,7 +425,6 @@
 			input = 0;
 		}
 
-
 		tuner.type = V4L2_TUNER_ANALOG_TV;
 		tuner.audmode = V4L2_TUNER_MODE_STEREO;
 		tuner.rxsubchans = V4L2_TUNER_SUB_STEREO;
@@ -440,6 +437,8 @@
 
 
 
+
+
 // Set norm
 		v4l2_std_id std_id;
 		switch(device->channel->norm)
