--- cinecutie/vdevicex11.C	2008-07-21 21:14:00.000000000 +0200
+++ /home/user/cinecutie/cinecutie/vdevicex11.C	2010-02-19 22:19:12.018571035 +0100
@@ -352,7 +352,6 @@
 			{
 				int size_change = (bitmap->get_w() != output->get_canvas()->get_w() ||
 					bitmap->get_h() != output->get_canvas()->get_h());
-printf("VDeviceX11::new_output_buffer 1\n");
 				delete bitmap;
 				delete output_frame;
 				bitmap = 0;
@@ -489,6 +488,12 @@
 					device->out_w,
 					device->out_h,
 					colormodel);
+// printf("VDeviceX11::new_outout_buffer %p %d %d %d %p\n", 
+// device,
+// device->out_w,
+// device->out_h,
+// colormodel,
+// output_frame->get_rows());
 //BUFFER2(output_frame->get_rows()[0], "VDeviceX11::new_output_buffer 2");
 				bitmap_type = BITMAP_TEMP;
 			}
@@ -508,11 +513,12 @@
 		}
 	}
 
+
 	*result = output_frame;
-//printf("VDeviceX11::new_output_buffer 10 %d\n", output->get_canvas()->get_window_lock());
 
 	output->get_canvas()->unlock_window();
 	output->unlock_canvas();
+//printf("VDeviceX11::new_output_buffer 10\n");
 }
 
 
@@ -535,12 +541,16 @@
 
 int VDeviceX11::write_buffer(VFrame *output_channels, EDL *edl)
 {
+// The reason for not drawing single frame is that it is _always_ drawn 
+// when drawing draw_refresh in cwindowgui and vwindowgui
+	if (device->single_frame) 
+		return 0;
+
 	int i = 0;
 	output->lock_canvas("VDeviceX11::write_buffer");
 	output->get_canvas()->lock_window("VDeviceX11::write_buffer 1");
 
 
-
 //printf("VDeviceX11::write_buffer %d\n", output->get_canvas()->get_video_on());
 	output->get_transfers(edl, 
 		output_x1, 
