--- cinecutie/viewmenu.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/viewmenu.C	2010-02-19 22:19:12.018571035 +0100
@@ -35,7 +35,7 @@
 
 
 
-ShowAssets::ShowAssets(MWindow *mwindow, const char *hotkey)
+ShowAssets::ShowAssets(MWindow *mwindow, char *hotkey)
  : BC_MenuItem(_("Show assets"), hotkey, hotkey[0])
 {
 	this->mwindow = mwindow; 
@@ -62,7 +62,7 @@
 
 
 
-ShowTitles::ShowTitles(MWindow *mwindow, const char *hotkey)
+ShowTitles::ShowTitles(MWindow *mwindow, char *hotkey)
  : BC_MenuItem(_("Show titles"), hotkey, hotkey[0])
 {
 	this->mwindow = mwindow; 
@@ -88,7 +88,7 @@
 
 
 
-ShowTransitions::ShowTransitions(MWindow *mwindow, const char *hotkey)
+ShowTransitions::ShowTransitions(MWindow *mwindow, char *hotkey)
  : BC_MenuItem(_("Show transitions"), hotkey, hotkey[0])
 { 
 	this->mwindow = mwindow; 
@@ -112,8 +112,8 @@
 
 
 ShowAutomation::ShowAutomation(MWindow *mwindow, 
-	const char *text,
-	const char *hotkey,
+	char *text,
+	char *hotkey,
 	int subscript)
  : BC_MenuItem(text, hotkey, hotkey[0])
 {
@@ -142,8 +142,8 @@
 
 
 
-PluginAutomation::PluginAutomation(MWindow *mwindow, const char *hotkey)
- : BC_MenuItem(_("Plugin keyframes"), hotkey, hotkey[0]) 
+PluginAutomation::PluginAutomation(MWindow *mwindow, char *hotkey)
+ : BC_MenuItem(_("Plugin Autos"), hotkey, hotkey[0]) 
 { 
 	this->mwindow = mwindow; 
 }
@@ -154,6 +154,7 @@
 	mwindow->edl->session->auto_conf->plugins = get_checked();
 	mwindow->gui->canvas->draw_overlays();
 	mwindow->gui->canvas->flash();
+//	mwindow->gui->mainmenu->draw_items();
 	mwindow->gui->unlock_window();
 	mwindow->gwindow->gui->update_toggles(1);
 	mwindow->gui->lock_window("PluginAutomation::handle_event");
