--- cinecutie/virtualanode.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/virtualanode.C	2010-02-19 22:19:12.018571035 +0100
@@ -25,7 +25,6 @@
 #include "arender.h"
 #include "atrack.h"
 #include "automation.h"
-#include "bcsignals.h"
 #include "edits.h"
 #include "edl.h"
 #include "edlsession.h"
@@ -229,7 +228,6 @@
 			start_position,
 			len,
 			sample_rate);
-
 	}
 	else
 // Read data from previous entity
