--- cinecutie/virtualnode.C	2008-07-26 03:56:57.000000000 +0200
+++ /home/user/cinecutie/cinecutie/virtualnode.C	2010-02-19 22:19:12.018571035 +0100
@@ -23,7 +23,6 @@
 #include "auto.h"
 #include "automation.h"
 #include "autos.h"
-#include "bcsignals.h"
 #include "floatauto.h"
 #include "floatautos.h"
 #include "intauto.h"
@@ -129,54 +128,9 @@
 			0,
 			1);
 
-		int circular_reference = 0;
 // Switch off if circular reference.  This happens if a plugin set or a track is deleted.
-		if(plugin == real_plugin) circular_reference = 1;
-
-// Switch off consecutive references to the same plugin
-		if(plugin &&
-			(plugin->plugin_type == PLUGIN_SHAREDPLUGIN ||
-			plugin->plugin_type == PLUGIN_SHAREDMODULE))
-		{
-			int real_module_number = plugin->shared_location.module;
-			int real_plugin_number = plugin->shared_location.plugin;
-
-			for(int j = i - 1; j >= 0; j--)
-			{
-				Plugin *prev_plugin = track->get_current_plugin(current_position, 
-					j, 
-					renderengine->command->get_direction(),
-					0,
-					1);
-				if(prev_plugin && 
-					prev_plugin->plugin_type == PLUGIN_SHAREDPLUGIN &&
-					plugin->plugin_type == PLUGIN_SHAREDPLUGIN)
-				{
-					int prev_module_number = prev_plugin->shared_location.module;
-					int prev_plugin_number = prev_plugin->shared_location.plugin;
-					if(real_module_number == prev_module_number &&
-						real_plugin_number == prev_plugin_number)
-					{
-						circular_reference = 1;
-						break;
-					}
-				}
-				else
-				if(prev_plugin && 
-					prev_plugin->plugin_type == PLUGIN_SHAREDMODULE &&
-					plugin->plugin_type == PLUGIN_SHAREDMODULE)
-				{
-					int prev_module_number = prev_plugin->shared_location.module;
-					if(real_module_number == prev_module_number)
-					{
-						circular_reference = 1;
-						break;
-					}
-				}
-			}
-		}
+		if(plugin == real_plugin) continue;
 
-		if(circular_reference) continue;
 
 		if(plugin && plugin->on)
 		{
@@ -280,6 +234,9 @@
 	{
 		int real_module_number = plugin->shared_location.module;
 		Module *real_module = vconsole->module_number(real_module_number);
+// If a track is deleted real_module is not found
+		if(!real_module) return 1;
+
 		Track *track = real_module->track;
 
 // Switch off if circular reference.  This happens if a track is deleted.
