--- cinecutie/vtrack.C	2008-07-21 21:14:01.000000000 +0200
+++ /home/user/cinecutie/cinecutie/vtrack.C	2010-02-19 22:19:12.028571315 +0100
@@ -59,12 +59,13 @@
 {
 }
 
-void VTrack::create_objects()
+int VTrack::create_objects()
 {
 	Track::create_objects();
 	automation = new VAutomation(edl, this);
 	automation->create_objects();
 	edits = new VEdits(edl, this);
+	return 0;
 }
 
 // Used by PlaybackEngine
