--- cinecutie/vwindowgui.C	2008-10-11 17:46:25.000000000 +0200
+++ /home/user/cinecutie/cinecutie/vwindowgui.C	2010-02-19 22:19:12.028571315 +0100
@@ -70,6 +70,7 @@
 {
 	this->mwindow = mwindow;
 	this->vwindow = vwindow;
+	strcpy(loaded_title, "");
 }
 
 VWindowGUI::~VWindowGUI()
@@ -78,7 +79,7 @@
 	delete transport;
 }
 
-void VWindowGUI::change_source(EDL *edl, const char *title)
+void VWindowGUI::change_source(EDL *edl, char *title)
 {
 	update_sources(title);
 	char string[BCTEXTLEN];
@@ -86,7 +87,7 @@
 		sprintf(string, PROGRAM_NAME ": %s", title);
 	else
 		sprintf(string, PROGRAM_NAME);
-
+	strcpy(loaded_title, title);
 	lock_window("VWindowGUI::change_source");
 	slider->set_position();
 	timebar->update();
@@ -96,7 +97,7 @@
 
 
 // Get source list from master EDL
-void VWindowGUI::update_sources(const char *title)
+void VWindowGUI::update_sources(char *title)
 {
 	lock_window("VWindowGUI::update_sources");
 
@@ -158,11 +159,10 @@
 	unlock_window();
 }
 
-void VWindowGUI::create_objects()
+int VWindowGUI::create_objects()
 {
 	in_point = 0;
 	out_point = 0;
-	lock_window("VWindowGUI::create_objects");
 	set_icon(mwindow->theme->get_image("vwindow_icon"));
 
 //printf("VWindowGUI::create_objects 1\n");
@@ -233,7 +233,7 @@
 //printf("VWindowGUI::create_objects 2\n");
 	deactivate();
 	slider->activate();
-	unlock_window();
+	return 0;
 }
 
 int VWindowGUI::resize_event(int w, int h)
@@ -483,6 +483,7 @@
 		0,
 		0,
 		0,
+		0, // locklabels
 		1,
 		1,
 		1,
@@ -650,12 +651,23 @@
 		new_edl->create_objects();
 		new_edl->load_xml(mwindow->plugindb, &file, LOAD_ALL);
 		sprintf(new_edl->local_session->clip_title, _("Clip %d"), mwindow->session->clip_number++);
+		char string[BCTEXTLEN];
+		Units::totext(string, 
+				end - start, 
+				edl->session->time_format, 
+				edl->session->sample_rate, 
+				edl->session->frame_rate,
+				edl->session->frames_per_foot);
+
+		sprintf(new_edl->local_session->clip_notes, _("%s\n Created from:\n%s"), string, vwindow->gui->loaded_title);
+
 		new_edl->local_session->set_selectionstart(0);
 		new_edl->local_session->set_selectionend(0);
 
 
 //printf("VWindowEditing::to_clip 1 %s\n", edl->local_session->clip_title);
-
+		new_edl->local_session->set_selectionstart(0.0);
+		new_edl->local_session->set_selectionend(0.0);
 		vwindow->clip_edit->create_clip(new_edl);
 	}
 }
@@ -684,6 +696,7 @@
 	this->vwindow = vwindow;
 	this->gui = gui;
 	set_precision(0.00001);
+	set_pagination(1.0, 10.0);
 }
 
 VWindowSlider::~VWindowSlider()
